<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Information and simple tools related to genomics information">

<title>bio – metagentorch</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-0652833b129a4aecb7d8777fd89a11f4.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="bio – metagentorch">
<meta property="og:description" content="Information and simple tools related to genomics information">
<meta property="og:image" content="https://vtecftwy.github.io/metagentorch/03_bio_files/figure-html/cell-14-1-image.png">
<meta property="og:site_name" content="metagentorch">
<meta property="og:image:height" content="441">
<meta property="og:image:width" content="703">
<meta name="twitter:title" content="bio – metagentorch">
<meta name="twitter:description" content="Information and simple tools related to genomics information">
<meta name="twitter:image" content="https://vtecftwy.github.io/metagentorch/03_bio_files/figure-html/cell-14-1-image.png">
<meta name="twitter:image-height" content="441">
<meta name="twitter:image-width" content="703">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">metagentorch</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./core.html">General Code</a></li><li class="breadcrumb-item"><a href="./bio.html">bio</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">metagentorch</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">General Code</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">core</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./art.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">art</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bio.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">bio</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wandb.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">wandb</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">CNN Virus</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_architecture.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">architecture</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_utils.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">utils</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#information-on-sequence-file-formats" id="toc-information-on-sequence-file-formats" class="nav-link active" data-scroll-target="#information-on-sequence-file-formats">Information on sequence file formats</a>
  <ul class="collapse">
  <li><a href="#sequence-file-formats" id="toc-sequence-file-formats" class="nav-link" data-scroll-target="#sequence-file-formats">Sequence File Formats</a>
  <ul class="collapse">
  <li><a href="#fasta-format" id="toc-fasta-format" class="nav-link" data-scroll-target="#fasta-format">FASTA format</a></li>
  </ul></li>
  <li><a href="#read-files" id="toc-read-files" class="nav-link" data-scroll-target="#read-files">Read Files</a>
  <ul class="collapse">
  <li><a href="#fastq-format" id="toc-fastq-format" class="nav-link" data-scroll-target="#fastq-format">FASTQ format</a></li>
  <li><a href="#aln-alignment-file-format" id="toc-aln-alignment-file-format" class="nav-link" data-scroll-target="#aln-alignment-file-format">ALN alignment file format</a></li>
  <li><a href="#q_score2prob_error" id="toc-q_score2prob_error" class="nav-link" data-scroll-target="#q_score2prob_error">q_score2prob_error</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#standard-codon-tables" id="toc-standard-codon-tables" class="nav-link" data-scroll-target="#standard-codon-tables">Standard codon tables</a>
  <ul class="collapse">
  <li><a href="#standard-dna-codon-table" id="toc-standard-dna-codon-table" class="nav-link" data-scroll-target="#standard-dna-codon-table">Standard DNA codon table</a>
  <ul class="collapse">
  <li><a href="#standarddnacodon" id="toc-standarddnacodon" class="nav-link" data-scroll-target="#standarddnacodon">StandardDNACodon</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/vtecftwy/metagentorch/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./core.html">General Code</a></li><li class="breadcrumb-item"><a href="./bio.html">bio</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">bio</h1>
</div>

<div>
  <div class="description">
    Information and simple tools related to genomics information
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="information-on-sequence-file-formats" class="level1">
<h1>Information on sequence file formats</h1>
<section id="sequence-file-formats" class="level2">
<h2 class="anchored" data-anchor-id="sequence-file-formats">Sequence File Formats</h2>
<p>FASTA is a standard format to store genetic sequences.</p>
<section id="fasta-format" class="level3">
<h3 class="anchored" data-anchor-id="fasta-format">FASTA format</h3>
<p>FASTA files count two lines for each sequence (see reference on <a href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat/">NCBI site</a>):</p>
<ul>
<li>line 1: the fasta definition line starting with <code>&gt;</code> and a unique identifier, followed by optional additional information.</li>
<li>line 2: the sequence line: <code>TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCA ...</code></li>
</ul>
<p>The format of the definition line varies from data source to data source.</p>
<p>Example:</p>
<pre><code>    &gt;MCHU - Calmodulin - Human, rabbit, bovine, rat, and chicken
    MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTIDFP ....</code></pre>
</section>
</section>
<section id="read-files" class="level2">
<h2 class="anchored" data-anchor-id="read-files">Read Files</h2>
<p>In this project, we use read sequences from NextGen sequencing tools as well as read sequences simulated using the ART Illumina simulator tool.</p>
<p><em>NextGen sequencing tools</em> most often store read sequences in a <strong>FASTQ</strong> format.</p>
<p><em>ART Illumina simulator tool</em> outputs both a <strong>FASTQ</strong> file and an alignment file (<strong>ALN</strong>) providing information on the original reference sequence it was simulated from. ALN files are also referred to as Clustal Alignnent format.</p>
<section id="fastq-format" class="level3">
<h3 class="anchored" data-anchor-id="fastq-format">FASTQ format</h3>
<p>A FASTQ file contains one 4-line block for each read. It includes a readID, followed by both the read sequence and a quality scores, in the following format:</p>
<pre class="text"><code>    @readid 
    sequence_read 
    + 
    base_quality_scores </code></pre>
<p>Example:</p>
<pre class="text"><code>    @2591237:ncbi:1-60400
    ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
    +
    CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></pre>
<p>For each base in the read sequence, a base quality score is coded by an ASCII character. They are <em>Q Scores</em> or <em>Phred+33</em> encoded, where the quality score is equal to ASCII code of the character minus 33.</p>
<blockquote class="blockquote">
<p>Reference information:</p>
<ul>
<li><a href="https://gensoft.pasteur.fr/docs/ART/2016-06-05/art_illumina_README">ART Illumina README file</a></li>
<li><a href="https://support.illumina.com/bulletins/2016/04/fastq-files-explained.html">Illumina site</a></li>
</ul>
</blockquote>
<p><strong>Information extraction</strong></p>
<p>The following metadata information can be parsed from <strong>FASTQ</strong> generated by <strong>ART Illumina</strong>:</p>
<p>FASTQ file element:</p>
<pre class="text"><code>    @2591237:ncbi:1-60400
    ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
    +
    CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></pre>
<ul>
<li><code>readid</code>, e.g.&nbsp;<code>2591237:ncbi:1-60400</code>.</li>
<li><code>read_sequence</code> itself, e.g.&nbsp;<code>ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG</code></li>
<li><code>base_quality_score</code>, the Q Score in ASCII, e.g.&nbsp;<code>CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></li>
</ul>
<p>Furthermore, the following metadata can be parsed from the first line:</p>
<pre><code>- `refseqid` of the original sequence used for the read, e.g. `2591237:nbci-1`
- `read_nbr` of the read, e.g. `1-60400</code></pre>
</section>
<section id="aln-alignment-file-format" class="level3">
<h3 class="anchored" data-anchor-id="aln-alignment-file-format">ALN alignment file format</h3>
<p>An alignment file in ALN format has a <code>Header</code> and main <code>Body</code> parts.</p>
<p>The header part includes the command used to generate this file and reference sequence id and length. The header <code>@CM</code> tag for command line, and <code>@SQ</code> for reference sequence.</p>
<p>A <code>Header</code> always starts with <code>##ART</code> and ends with <code>##Header End</code>.</p>
<p><code>Header</code> example, for ALN file generated by ART Illumina:</p>
<pre class="ascii"><code>    ##ART_Illumina  read_length 50
    @CM /bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/cov_data/cov_virus_sequences_ten.fa -ss HS25 -l 50 -f 100 -o /home/vtec/projects/bio/metagentools/data/cov_simreads/single_10seq_50bp/single_10seq_50bp -rs 1674660835
    @SQ 2591237:ncbi:1 [MK211378]   2591237 ncbi    1 [MK211378] 2591237    Coronavirus BtRs-BetaCoV/YN2018D    30213
    @SQ 11128:ncbi:2 [LC494191] 11128   ncbi    2 [LC494191] 11128  Bovine coronavirus      30942
    @SQ 31631:ncbi:3 [KY967361] 31631   ncbi    3 [KY967361] 31631  Human coronavirus OC43      30661
    ##Header End</code></pre>
<p>For each reference sequence used to build the reads, we can parse the following metadata info:</p>
<pre class="ascii"><code>    refseqid:       2591237:ncbi:1
    ref_seq_length: 30213</code></pre>
<p>The <code>Body</code> part contains each read’s alignments in the following format:</p>
<pre class="ascii"><code>    &gt;refseqid   read_id   aln_start_pos    ref_seq_strand
    ref_seq_aligned
    read_seq_aligned </code></pre>
<ul>
<li><code>aln_start_pos</code> is the alignment start position of the read in the reference sequence.</li>
<li><code>ref_seq_aligned</code> is the aligned region of reference sequence, which can be from plus strand or minus strand of the reference sequence.<br>
</li>
<li><code>read_seq_aligned</code> is the aligned sequence read, which always in the same orientation of the same read in the corresponding fastq file.</li>
</ul>
<p><code>aln_start_pos</code> is always relative to the strand of reference sequence. That is, <code>aln_start_pos</code> 10 in the plus (+) strand is different from <code>aln_start_pos</code> 10 in the minus (‐) stand.</p>
<p><code>Body</code> example:</p>
<pre class="ascii"><code>    &gt;2591237:ncbi:1 2591237:ncbi:1-60400    14770   +
    ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
    ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG</code></pre>
<pre class="ascii"><code>    refseqid:          2591237:ncbi:1
    readid:            2591237:ncbi:1-20100
    aln_start_pos:     23878
    ref_seq_strand:    -
    ref_seq_aligned:   ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
    read_seq_aligned:  ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG</code></pre>
<blockquote class="blockquote">
<p>Reference information:</p>
<ul>
<li><a href="https://gensoft.pasteur.fr/docs/ART/2016-06-05/art_illumina_README">ART Illumina README file</a></li>
<li><a href="https://support.illumina.com/bulletins/2016/04/fastq-files-explained.html">Illumina site</a></li>
</ul>
</blockquote>
<p>The quality score of a base, also known as a <em>Phred</em> or <em>Q score</em>, is an integer value representing the estimated probability of an error, i.e.&nbsp;the probability that the base is incorrectly sequenced.</p>
<p>If <span class="math inline">\(P\)</span> is the error probability and <span class="math inline">\(Q\)</span> is the quality score for a base, then:</p>
<ul>
<li><span class="math inline">\(P = 10^{-Q/10}\)</span></li>
<li><span class="math inline">\(Q = -10 \log_{10}(P)\)</span></li>
</ul>
<p>Q scores are often represented as ASCII characters. The rule for converting an ASCII character to an integer varies.</p>
<p>The two tables below convert between integer Q scores, ASCII characters and error probabilities.</p>
<p><strong>ASCII_BASE = 33</strong> (table on top) is now almost universally used. In older systems, ASCII_BASE = 64 is also used.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="03_bio_files/figure-html/cell-14-1-image.png" class="img-fluid figure-img"></p>
<figcaption>image.png</figcaption>
</figure>
</div>
<p>(See reference <a href="http://drive5.com/usearch/manual/quality_score.html">here</a>)</p>
<section id="q-score-examples" class="level4">
<h4 class="anchored" data-anchor-id="q-score-examples">Q Score examples</h4>
<ul>
<li>ASCII character is <code>K</code></li>
<li>ASCII code is <code>75</code></li>
<li>Q score: $Q = 75 - $ <span class="math inline">\(= 75 - 33 = 42\)</span></li>
<li>P_error = <span class="math inline">\(P = 10^{-Q/10}\)</span>, i.e.&nbsp;<span class="math inline">\(10^{-42/10}\)</span> = 0.00006</li>
</ul>
<blockquote class="blockquote">
<p><strong>Python note</strong>:</p>
<p>Decode an ASCII character with <code>ord(c)</code> and encode it with <code>chr(i)</code></p>
</blockquote>
<div id="cell-17" class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="bu">ord</span>(<span class="st">'!'</span>), <span class="bu">chr</span>(<span class="dv">33</span>), <span class="bu">ord</span>(<span class="st">'K'</span>), <span class="bu">chr</span>(<span class="dv">75</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>(33, '!', 75, 'K')</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/bio.py#L9" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
</section>
<section id="q_score2prob_error" class="level3">
<h3 class="anchored" data-anchor-id="q_score2prob_error">q_score2prob_error</h3>
<blockquote class="blockquote">
<pre><code> q_score2prob_error (char:str, ASCII_base:int=33)</code></pre>
</blockquote>
<p><em>Return the probability of error for a given Q score encoded as ASCII character</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>char</td>
<td>str</td>
<td></td>
<td>ASCII character retrieved from Q Score or Phred value in FASTQ</td>
</tr>
<tr class="even">
<td>ASCII_base</td>
<td>int</td>
<td>33</td>
<td>ASCII base. Mostly 33, can be 64 in old FASTQ files</td>
</tr>
</tbody>
</table>
<div id="cell-19" class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Selection of high Q score ASICC code, Q score values and corresponding P error:
ASCII Q Score | Q Score | P error
--------------|---------|--------
      K       |   42    | 0.00006 
      J       |   41    | 0.00008 
      I       |   40    | 0.00010 
      H       |   39    | 0.00013 
      G       |   38    | 0.00016 
      F       |   37    | 0.00020 
      E       |   36    | 0.00025 
      D       |   35    | 0.00032 
      C       |   34    | 0.00040 
      B       |   33    | 0.00050 
      A       |   32    | 0.00063 
      @       |   31    | 0.00079 
      ?       |   30    | 0.00100 
      &gt;       |   29    | 0.00126 
      =       |   28    | 0.00158 </code></pre>
</div>
</div>
<p>In the case of the following FASTQ file:</p>
<pre class="ascii"><code>    @2591237:ncbi:1-60400
    ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
    +
    CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></pre>
<p>The following metadata information can be parsed:</p>
<ul>
<li><code>read_id</code>, e.g.&nbsp;<code>2591237:ncbi:1-60400</code>. The following information can be parsed:
<ul>
<li><code>ref_seq_id</code> of the original sequence used for the read, e.g.&nbsp;<code>2591237:nbci-1</code></li>
<li><code>read_nbr</code> of the read, e.g.&nbsp;`60400</li>
</ul></li>
<li><code>sequence_read</code> itself, e.g.&nbsp;<code>ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG</code></li>
<li><code>base_quality_score</code>, the Q Score in ASCII, e.g.&nbsp;<code>CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></li>
</ul>
<p>This ASCII encoded Q Score can be converted into probability of error for each bp in the read:</p>
<div id="cell-22" class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>q_scores <span class="op">=</span> <span class="st">'CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG'</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>num_scores <span class="op">=</span> [<span class="ss">f"</span><span class="sc">{</span>q_score2prob_error(char)<span class="sc">:6.5f}</span><span class="ss">"</span> <span class="cf">for</span> char <span class="kw">in</span> q_scores]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'; '</span>.join(num_scores))</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Maximum error probability: </span><span class="sc">{</span><span class="bu">max</span>(num_scores)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.00040; 0.00040; 0.00040; 0.00050; 0.00040; 0.00016; 0.00020; 0.00016; 0.00050; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00050; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00126; 0.00016; 0.00016; 0.00016; 0.02512; 0.00016; 0.00158; 0.03981; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016; 0.00016
Maximum error probability: 0.03981</code></pre>
</div>
</div>
<p>It shows that most bp have a very low probability of error, except a few bp with probability or 2.5% and 3.9%,</p>
</section>
</section>
</section>
<section id="standard-codon-tables" class="level1">
<h1>Standard codon tables</h1>
<section id="standard-dna-codon-table" class="level2">
<h2 class="anchored" data-anchor-id="standard-dna-codon-table">Standard DNA codon table</h2>
<p>These are the basic codon tables for DNA sequences. They are used to translate DNA sequences into amino acid sequences.</p>
<p><strong>Direct DNA codon table</strong></p>
<p><img src="03_bio_files/figure-html/cell-28-1-image.png" class="img-fluid"></p>
<p><strong>Inverse DNA codon table</strong></p>
<p><img src="03_bio_files/figure-html/cell-30-1-image.png" class="img-fluid"></p>
<p>(<a href="https://en.wikipedia.org/wiki/DNA_and_RNA_codon_tables">Source</a>)</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/bio.py#L20" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="standarddnacodon" class="level3">
<h3 class="anchored" data-anchor-id="standarddnacodon">StandardDNACodon</h3>
<blockquote class="blockquote">
<pre><code> StandardDNACodon ()</code></pre>
</blockquote>
<p>*Hold standard DNA codon reference information</p>
<ol type="1">
<li>DNA codon direct table: codon -&gt; amino acid information</li>
<li>DNA codin inverse table amino acid code (e.g.&nbsp;Cys) -&gt; list of codons*</li>
</ol>
<div id="cell-33" class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>codons <span class="op">=</span> StandardDNACodon()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-34" class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>key <span class="op">=</span> <span class="st">'Phe'</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>codons.inverse_table[key]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'amino acid name': 'Phenylalanine',
 'amino acid letter': 'F',
 'codons': ['TTT', 'TTC']}</code></pre>
</div>
</div>
<div id="cell-35" class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>key <span class="op">=</span> <span class="st">'TTT'</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>codons.direct_table[key]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'amino acid symbol': 'Phe',
 'amino acid name': 'Phenylalanine',
 'amino acid letter': 'F'}</code></pre>
</div>
</div>
<div id="cell-36" class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>codons.amino_acid_symbols[:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['Ala', 'Arg', 'Asn', 'Asp', 'Cys']</code></pre>
</div>
</div>
<div id="cell-37" class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>codons.amino_acid_names[:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['Alanine', 'Arginine', 'Asparagine', 'Aspartic acid', 'Cysteine']</code></pre>
</div>
</div>
<div id="cell-38" class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>codons.amino_acid_letters[:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['A', 'R', 'N', 'D', 'C']</code></pre>
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/vtecftwy\.github\.io\/metagentorch");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/vtecftwy/metagentorch/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>