<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Data preprocessing and transform tools for CNN Virus data.">

<title>data – metagentorch</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-0652833b129a4aecb7d8777fd89a11f4.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="data – metagentorch">
<meta property="og:description" content="Data preprocessing and transform tools for CNN Virus data.">
<meta property="og:site_name" content="metagentorch">
<meta name="twitter:title" content="data – metagentorch">
<meta name="twitter:description" content="Data preprocessing and transform tools for CNN Virus data.">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">metagentorch</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./cnn_virus_architecture.html">CNN Virus</a></li><li class="breadcrumb-item"><a href="./cnn_virus_data.html">data</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">metagentorch</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">General Code</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">core</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./art.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">art</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">bio</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wandb.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">wandb</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">CNN Virus</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_architecture.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">architecture</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_data.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cnn_virus_utils.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">utils</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#cnn-virus-project-data-structure" id="toc-cnn-virus-project-data-structure" class="nav-link active" data-scroll-target="#cnn-virus-project-data-structure">CNN Virus project data structure</a>
  <ul class="collapse">
  <li><a href="#data-directory" id="toc-data-directory" class="nav-link" data-scroll-target="#data-directory">Data directory</a></li>
  <li><a href="#original-datasets" id="toc-original-datasets" class="nav-link" data-scroll-target="#original-datasets">Original datasets</a>
  <ul class="collapse">
  <li><a href="#originallabels" id="toc-originallabels" class="nav-link" data-scroll-target="#originallabels">OriginalLabels</a></li>
  <li><a href="#originallabels.search" id="toc-originallabels.search" class="nav-link" data-scroll-target="#originallabels.search">OriginalLabels.search</a></li>
  </ul></li>
  <li><a href="#ncbi-reference-sequences" id="toc-ncbi-reference-sequences" class="nav-link" data-scroll-target="#ncbi-reference-sequences">NCBI reference sequences</a></li>
  <li><a href="#model-related-data" id="toc-model-related-data" class="nav-link" data-scroll-target="#model-related-data">Model related data</a></li>
  </ul></li>
  <li><a href="#parsing-sequence-files" id="toc-parsing-sequence-files" class="nav-link" data-scroll-target="#parsing-sequence-files">Parsing sequence files</a>
  <ul class="collapse">
  <li><a href="#fasta-file" id="toc-fasta-file" class="nav-link" data-scroll-target="#fasta-file">FASTA file</a>
  <ul class="collapse">
  <li><a href="#fastafilereader" id="toc-fastafilereader" class="nav-link" data-scroll-target="#fastafilereader">FastaFileReader</a></li>
  <li><a href="#fastafilereader.review" id="toc-fastafilereader.review" class="nav-link" data-scroll-target="#fastafilereader.review">FastaFileReader.review</a></li>
  <li><a href="#fastafilereader.print_first_chunks" id="toc-fastafilereader.print_first_chunks" class="nav-link" data-scroll-target="#fastafilereader.print_first_chunks">FastaFileReader.print_first_chunks</a></li>
  <li><a href="#parsing-metadata" id="toc-parsing-metadata" class="nav-link" data-scroll-target="#parsing-metadata">Parsing metadata</a></li>
  <li><a href="#textfilebasereader.parse_text" id="toc-textfilebasereader.parse_text" class="nav-link" data-scroll-target="#textfilebasereader.parse_text">TextFileBaseReader.parse_text</a></li>
  <li><a href="#textfilebasereader.set_parsing_rules" id="toc-textfilebasereader.set_parsing_rules" class="nav-link" data-scroll-target="#textfilebasereader.set_parsing_rules">TextFileBaseReader.set_parsing_rules</a></li>
  <li><a href="#fastafilereader.parse_file" id="toc-fastafilereader.parse_file" class="nav-link" data-scroll-target="#fastafilereader.parse_file">FastaFileReader.parse_file</a></li>
  </ul></li>
  <li><a href="#fastq-file" id="toc-fastq-file" class="nav-link" data-scroll-target="#fastq-file">FASTQ file</a>
  <ul class="collapse">
  <li><a href="#fastqfilereader" id="toc-fastqfilereader" class="nav-link" data-scroll-target="#fastqfilereader">FastqFileReader</a></li>
  <li><a href="#fastqfilereader.parse_file" id="toc-fastqfilereader.parse_file" class="nav-link" data-scroll-target="#fastqfilereader.parse_file">FastqFileReader.parse_file</a></li>
  </ul></li>
  <li><a href="#aln-alignment-files" id="toc-aln-alignment-files" class="nav-link" data-scroll-target="#aln-alignment-files">ALN Alignment Files</a>
  <ul class="collapse">
  <li><a href="#alnfilereader" id="toc-alnfilereader" class="nav-link" data-scroll-target="#alnfilereader">AlnFileReader</a></li>
  <li><a href="#textfilebasereader.parse_text-1" id="toc-textfilebasereader.parse_text-1" class="nav-link" data-scroll-target="#textfilebasereader.parse_text-1">TextFileBaseReader.parse_text</a></li>
  <li><a href="#alnfilereader.parse_definition_line_with_position" id="toc-alnfilereader.parse_definition_line_with_position" class="nav-link" data-scroll-target="#alnfilereader.parse_definition_line_with_position">AlnFileReader.parse_definition_line_with_position</a></li>
  <li><a href="#alnfilereader.parse_file" id="toc-alnfilereader.parse_file" class="nav-link" data-scroll-target="#alnfilereader.parse_file">AlnFileReader.parse_file</a></li>
  <li><a href="#alnfilereader.parse_header_reference_sequences" id="toc-alnfilereader.parse_header_reference_sequences" class="nav-link" data-scroll-target="#alnfilereader.parse_header_reference_sequences">AlnFileReader.parse_header_reference_sequences</a></li>
  <li><a href="#alnfilereader.set_header_parsing_rules" id="toc-alnfilereader.set_header_parsing_rules" class="nav-link" data-scroll-target="#alnfilereader.set_header_parsing_rules">AlnFileReader.set_header_parsing_rules</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#build-datasets" id="toc-build-datasets" class="nav-link" data-scroll-target="#build-datasets">Build datasets</a>
  <ul class="collapse">
  <li><a href="#plain-text-based-data" id="toc-plain-text-based-data" class="nav-link" data-scroll-target="#plain-text-based-data">Plain text based data</a>
  <ul class="collapse">
  <li><a href="#textfiledataset" id="toc-textfiledataset" class="nav-link" data-scroll-target="#textfiledataset">TextFileDataset</a></li>
  </ul></li>
  <li><a href="#aln-file-based" id="toc-aln-file-based" class="nav-link" data-scroll-target="#aln-file-based">ALN file based</a>
  <ul class="collapse">
  <li><a href="#alnfiledataset" id="toc-alnfiledataset" class="nav-link" data-scroll-target="#alnfiledataset">AlnFileDataset</a></li>
  </ul></li>
  <li><a href="#handle-long-reads" id="toc-handle-long-reads" class="nav-link" data-scroll-target="#handle-long-reads">Handle long reads</a>
  <ul class="collapse">
  <li><a href="#preprocessing" id="toc-preprocessing" class="nav-link" data-scroll-target="#preprocessing">Preprocessing</a></li>
  <li><a href="#split_kmer_batch_into_50mers" id="toc-split_kmer_batch_into_50mers" class="nav-link" data-scroll-target="#split_kmer_batch_into_50mers">split_kmer_batch_into_50mers</a></li>
  <li><a href="#postprocessing" id="toc-postprocessing" class="nav-link" data-scroll-target="#postprocessing">Postprocessing</a></li>
  <li><a href="#combine_predictions" id="toc-combine_predictions" class="nav-link" data-scroll-target="#combine_predictions">combine_predictions</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#deprecated-items" id="toc-deprecated-items" class="nav-link" data-scroll-target="#deprecated-items">Deprecated Items</a>
  <ul class="collapse">
  <li><a href="#combine_prediction_batch" id="toc-combine_prediction_batch" class="nav-link" data-scroll-target="#combine_prediction_batch">combine_prediction_batch</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/vtecftwy/metagentorch/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./cnn_virus_architecture.html">CNN Virus</a></li><li class="breadcrumb-item"><a href="./cnn_virus_data.html">data</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">data</h1>
</div>

<div>
  <div class="description">
    Data preprocessing and transform tools for CNN Virus data.
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="cnn-virus-project-data-structure" class="level1">
<h1>CNN Virus project data structure</h1>
<p>There are many different types of files and datasets for this project. All data are located in directory <code>data</code>, under the project root. The following is an overview of the main types of data and where they sit in the directory tree.</p>
<section id="data-directory" class="level2">
<h2 class="anchored" data-anchor-id="data-directory">Data directory</h2>
<p>Let’s create an instance of <a href="https://vtecftwy.github.io/metagentorch/core.html#projectfilesystem"><code>ProjectFileSystem</code></a> to access the data directory and its content.</p>
<div id="cell-6" class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>pfs <span class="op">=</span> ProjectFileSystem()</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>pfs.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Running linux on local computer
Device's home directory: /home/vtec
Project file structure:
 - Root ........ /home/vtec/projects/bio/metagentorch 
 - Data Dir .... /home/vtec/projects/bio/metagentorch/nbs-dev/data_dev 
 - Notebooks ... /home/vtec/projects/bio/metagentorch/nbs</code></pre>
</div>
</div>
<p>Note that we have setup this notebook to use the development data directory <code>metagentorch/nbs-dev/data_dev</code> and not the standard <code>metagentorch/data</code>.</p>
<p>In each directory, a <code>readme.md</code> file or another <code>*.md</code> file can be added to provide a description of the directory content.</p>
<p>These <code>readme.md</code> files can be conveniently accessed using the <code>.readme(path)</code> method available with the class <a href="https://vtecftwy.github.io/metagentorch/core.html#projectfilesystem"><code>ProjectFileSystem</code></a> (from core module).</p>
<div id="cell-9" class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>pfs.readme()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="data-directory-for-this-package-development" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="data-directory-for-this-package-development">Data directory for this package development</h3>
<p>This directory includes all data required to validate and test this package code.</p>
<pre class="text"><code>data_dev
 |--- CNN_Virus_data
 |     |--- 50mer_ds_100_seq
 |     |--- 150mer_ds_100_seq
 |     |--- train_short
 |     |--- val_short
 |     |--- weight_of_classes
 |--- ncbi
 |     |--- infer_results
 |     |     |--- cnn_virus
 |     |     |--- csv
 |     |     |--- xlsx
 |     |     |--- testdb.db
 |     |--- refsequences
 |     |     |--- cov
 |     |     |     |--cov_virus_sequence_one_metadata.json
 |     |     |     |--sequences_two_no_matching_rule.fa
 |     |     |     |--another_sequence.fa
 |     |     |     |--cov_virus_sequences_two.fa
 |     |     |     |--cov_virus_sequences_two_metadata.json
 |     |     |     |--cov_virus_sequence_one.fa
 |     |     |     |--single_1seq_150bp
 |     |     |     |    |--single_1seq_150bp.fq
 |     |     |     |    |--single_1seq_150bp.aln
 |     |     |     |--paired_1seq_150bp
 |     |     |     |    |--paired_1seq_150bp2.aln
 |     |     |     |    |--paired_1seq_150bp2.fq
 |     |     |     |    |--paired_1seq_150bp1.fq 
 |     |     |     |    |--paired_1seq_150bp1.aln 
 |     |--- simreads
 |     |     |--- cov
 |     |     |     |--- paired_1seq_50bp
 |     |     |     |      |--- paired_1seq_50bp_1.aln
 |     |     |     |      |--- paired_1seq_50bp_1.fq
 |     |     |     |--- single_1seq_50bp
 |     |     |     |      |--- single_1seq_50bp_1.aln
 |     |     |     |      |--- single_1seq_50bp_1.fq
 |     |     |--- cov
 |     |     |     |--single_1seq_50bp
 |     |     |     |    |--single_1seq_50bp.aln
 |     |     |     |    |--single_1seq_50bp.fq
 |     |     |     |--single_1seq_150bp
 |     |     |     |    |--single_1seq_150bp.fq
 |     |     |     |    |--single_1seq_150bp.aln
 |     |     |     |--paired_1seq_150bp
 |     |     |     |    |--paired_1seq_150bp2.aln
 |     |     |     |    |--paired_1seq_150bp2.fq
 |     |     |     |    |--paired_1seq_150bp1.fq
 |     |     |     |    |--paired_1seq_150bp1.aln
 |--- saved           
 |--- readme.md               </code></pre>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
</section>
<section id="original-datasets" class="level2">
<h2 class="anchored" data-anchor-id="original-datasets">Original datasets</h2>
<p>The CNN Virus team provides training and validation/test datasets for their model. These datasets and the pretrained model parameters are are available on their Google drive shared directory <a href="https://drive.google.com/open?id=1sj0-NCSKjLta_Geg6EMo26rChmtWcOiI">here</a>. A set of shortened datasets are also available in the <code>data-dev</code> development data directory for testing the code.</p>
<div id="cell-12" class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data<span class="op">/</span><span class="st">'CNN_Virus_data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/CNN_Virus_data</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="cnn-virus-data-development-directory-version" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="cnn-virus-data-development-directory-version">CNN Virus data (development directory version)</h3>
<p>This directory includes a set of short data files used to test train and inference with the CNN Virus.</p>
<section id="file-list-and-description" class="level4">
<h4 class="anchored" data-anchor-id="file-list-and-description">File list and description:</h4>
<section id="mer" class="level5">
<h5 class="anchored" data-anchor-id="mer">50-mer</h5>
<p>50-mer reads and their labels, in <em>text format</em> with one line per sample. Each line consists of three components, separated by tabs: the 50-mer read or sequence, the virus species label and the position label:</p>
<pre class="text"><code>'TTACNAGCTCCAGTCTAAGATTGTAACTGGCCTTTTTAAAGATTGCTCTA    94    5\n'</code></pre>
<p>Files:</p>
<ul>
<li><code>50mer_ds_100_seq</code>: small dataset with 100 reads</li>
<li><code>5train_short</code>: small 1000-read subset from the original training dataset for experiments</li>
<li><code>val_short</code>: small 500-read subset from the original validation dataset for experiments</li>
</ul>
</section>
<section id="mer-1" class="level5">
<h5 class="anchored" data-anchor-id="mer-1">150-mer</h5>
<p>150-mer reads and their labels in <em>text format</em> in a similar format as above:</p>
<pre class="text"><code>'TTCTTTCACCACCACAACCAGTCGGCCGTGGAGAGGCGTCGCCGCGTCTCGTTCGTCGAGGCCGATCGACTGCCGCATGAGAGCGGGTGGTATTCTTCCGAAGACGACGGAGACCGGGACGGTGATGAGGAAACTGGAGAGAGCCACAAC    6    0\n'</code></pre>
<p>Files:</p>
<ul>
<li><code>150mer_ds_100_reads</code>: small subset of 100 reads from original <code>ICTV_150mer_benchmarking</code> file</li>
</ul>
</section>
<section id="other-files" class="level5">
<h5 class="anchored" data-anchor-id="other-files">Other files:</h5>
<ul>
<li><code>virus_name_mapping</code>: mapping between virus species and their numerical label</li>
<li><code>weight_of_classes</code>: weights for each virus species class in the training dataset</li>
</ul>
</section>
</section>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L43" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="originallabels" class="level3">
<h3 class="anchored" data-anchor-id="originallabels">OriginalLabels</h3>
<blockquote class="blockquote">
<pre><code> OriginalLabels (p2mapping:pathlib.Path|None=None)</code></pre>
</blockquote>
<p><em>Converts between labels and species name as per original training dataset</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p2mapping</td>
<td>pathlib.Path | None</td>
<td>None</td>
<td>Path to the mapping file. Uses <code>virus_name_mapping</code> by default</td>
</tr>
</tbody>
</table>
<p>Original data include 187 viruses, with label from 0 to 186.</p>
<p>With the class method <code>.label2species(n)</code> and <code>.species2label(species)</code> we can convert between the label and the species name.</p>
<div id="cell-15" class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>species <span class="op">=</span> OriginalLabels()</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> n <span class="kw">in</span> [<span class="dv">0</span>, <span class="dv">94</span>, <span class="dv">117</span>, <span class="dv">118</span>]:</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>n<span class="sc">:3d}</span><span class="ss"> -&gt; </span><span class="sc">{</span>species<span class="sc">.</span>label2species(n)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  0 -&gt; Variola_virus
 94 -&gt; Middle_East_respiratory_syndrome-related_coronavirus
117 -&gt; Severe_acute_respiratory_syndrome-related_coronavirus
118 -&gt; Yellow_fever_virus</code></pre>
</div>
</div>
<div id="cell-16" class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> s <span class="kw">in</span> [<span class="st">'Variola_virus'</span>, <span class="st">'Yellow_fever_virus'</span>]:</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>s<span class="sc">:20s}</span><span class="ss"> -&gt; </span><span class="sc">{</span>species<span class="sc">.</span>species2label(s)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Variola_virus        -&gt; 0
Yellow_fever_virus   -&gt; 118</code></pre>
</div>
</div>
<p>When looking for a numerical specie label it is often more convenient to use a partial name, and the method <code>.search(species)</code> because we do not need to know the full specie name.</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L60" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="originallabels.search" class="level3">
<h3 class="anchored" data-anchor-id="originallabels.search">OriginalLabels.search</h3>
<blockquote class="blockquote">
<pre><code> OriginalLabels.search (s:str)</code></pre>
</blockquote>
<p><em>Prints all species whose name contains the passed string, with their numerical label</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>s</td>
<td>str</td>
<td>string to search through all original virus species</td>
</tr>
</tbody>
</table>
<div id="cell-19" class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>species.search(<span class="st">'fever'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sandfly_fever_Naples_phlebovirus. Label: 35
Crimean-Congo_hemorrhagic_fever_orthonairovirus. Label: 76
Yellow_fever_virus. Label: 118
Rift_Valley_fever_phlebovirus. Label: 156</code></pre>
</div>
</div>
</section>
</section>
<section id="ncbi-reference-sequences" class="level2">
<h2 class="anchored" data-anchor-id="ncbi-reference-sequences">NCBI reference sequences</h2>
<p>We simulate reads using many reference sequences from the NCBI GenBank database. We group all reference sequences as well as all reads simulated from these reference sequences under the <code>ncbi</code> directory.</p>
<div id="cell-22" class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>pfs <span class="op">=</span> ProjectFileSystem()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-23" class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data <span class="op">/</span> <span class="st">'ncbi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/ncbi</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="ncbi-data" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="ncbi-data">NCBI Data</h3>
<p>This directory includes all data related to the work done with reference sequences from NCBI.</p>
<p>The data is organized in the following subfolders:</p>
<ul>
<li><code>refsequences</code>: reference CoV sequences downloaded from NCBI, and related metadata</li>
<li><code>simreads</code>: all data from simulated reads, using ART Illumina simulator and the reference sequences</li>
<li><code>infer_results</code>: results from the inference using models with the simulated reads</li>
<li><code>ds</code>: datasets in proper format for training or inference/prediction using the CNN Virus model</li>
</ul>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
<div id="cell-24" class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/ncbi/refsequences</code>:</p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>No markdown file in this folder</code></pre>
</div>
</div>
<div id="cell-25" class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/ncbi/simreads</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="ncbi-simulated-reads" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="ncbi-simulated-reads">NCBI simulated reads</h3>
<p>This directory includes all sets of simulated read sequence files generated from NCBI viral sequences using ARC Illumina.</p>
<pre class="ascii"><code>this-directory
    |--cov
    |    |
    |    |--single_10seq_50bp
    |    |    |--single_10seq_50bp.fq
    |    |    |--single_10seq_50bp.alnEnd
    |    |-- ...
    |    |--single_100seq_150bp
    |    |    |--single_100seq_150bp.fq
    |    |    |--single_100seq_150bp.aln
    |    |--paired_100seq_50bp
    |    |    |--paired_100seq_50bp2.aln
    |    |    |--paired_100seq_50bp1.aln
    |    |    |--paired_100seq_50bp2.fq
    |    |    |--paired_100seq_50bp1.fq
    |    |-- ...
    |    |
    |---yf
    |    |
    |    |--yf_AY968064-single-150bp
    |    |    |--yf_AY968064-single-1seq-150bp.fq
    |    |    |--yf_AY968064-single-1seq-150bp.aln
    |    |
    |--mRhiFer1
    |    |--mRhiFer1_v1.p.dna_rm.primary_assembly.1
    |    |    |--mRhiFer1_v1.p.dna_rm.primary_assembly.1.fq
    |    |    |--mRhiFer1_v1.p.dna_rm.primary_assembly.1.aln
    |    |
</code></pre>
<p>This directory includes several subdirectories, each for one virus, e.g.&nbsp;<code>cov</code> for corona, <code>yf</code> for yellow fever.</p>
<p>In each virus subdirectory, several simreads directory includes simulated reads with various parameters, named as <code>&lt;method&gt;_&lt;nb-seq&gt;_&lt;nb-bp&gt;</code> where” - <code>&lt;method&gt;</code> is either <code>single</code> or <code>paired</code> depending on the simulation method - <code>&lt;nb-seq&gt;</code> is the number of reference sequences used for simulation, and refers to the <code>fa</code> file used - <code>&lt;nb-bp&gt;</code> is the number of base pairs used to simulate reads</p>
<p>Each sub-directory includes simreads files made using a simulation method and a specific number of reference sequences. - <code>xxx.fq</code> and <code>xxx.aln</code> files when method is <code>single</code> - <code>xxx1.fq</code>, <code>xxx2.fq</code>, <code>xxx1.aln</code> and <code>xxx2.aln</code> files when method is <code>paired</code>.</p>
<p>Example: - <code>paired_10seq_50bp</code> means that the simreads were generated by using the <code>paired</code> method to simulate 50-bp reads, and using the <code>fa</code> file <code>cov_virus_sequences_010-seqs.fa</code>. - <code>single_100seq_50bp</code> means that the simreads were generated by using the <code>single</code> method to simulate 50-bp reads, and using the <code>fa</code> file <code>cov_virus_sequences_100-seqs.fa</code>. Note that this generated 20,660,104 reads !</p>
<section id="simread-file-formats" class="level4">
<h4 class="anchored" data-anchor-id="simread-file-formats">Simread file formats</h4>
<p>Simulated reads information is split between two files: - <strong>FASTQ</strong> (<code>.fq</code>) files providing the read sequences and their ASCII quality scores - <strong>ALN</strong> (<code>.aln</code>) files with alignment information</p>
<section id="fastq-.fq" class="level5">
<h5 class="anchored" data-anchor-id="fastq-.fq">FASTQ (<code>.fq</code>)</h5>
<p>FASTQ files generated by ART Illumina have the following structure (showing 5 reads), with 4 lines for each read:</p>
<pre class="ascii"><code>@2591237:ncbi:1-60400
ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
+
CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG
@2591237:ncbi:1-60399
GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC
+
BCBCCFGGGGGGGG1CGGGG&lt;GGBGGGGGFGCGGGGGGDGGG/GG1GGGG
@2591237:ncbi:1-60398
ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC
+
CCCCCGGGEGG1GGF1G/GGEGGGGGGGGGGGGFFGGGGGGGGGGDGGDG
@2591237:ncbi:1-60397
CGTAAAGTAGAGGCTGTATGGTAGCTAGCACAAATGCCAGCACCAATAGG
+
BCCCCGGGFGGGGGGFGGGGFGG1GGGGGGG&gt;GG1GGGGGGGGGGE&lt;GGG
@2591237:ncbi:1-60396
GGTATCGGGTATCTCCTGCATCAATGCAAGGTCTTACAAAGATAAATACT
+
CBCCCGGG@CGGGGGGGGGGGG=GFGGGGDGGGFG1GGGGGGGG@GGGGG</code></pre>
<p>The following information can be parsed from the each read sequence in the FASTQ file:</p>
<ul>
<li>Line 1: <code>readid</code>, a unique ID for the read, under for format <code>@readid</code></li>
<li>Line 2: <code>readseq</code>, the sequence of the read</li>
<li>Line 3: a separator <code>+</code></li>
<li>Line 4: <code>read_qscores</code>, the base quality scores encoded in ASCII</li>
</ul>
<p>Example:</p>
<pre><code>@2591237:ncbi:1-60400
ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
+
CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></pre>
<ul>
<li><code>readid</code> = <code>2591237:ncbi:1-60400</code></li>
<li><code>readseq</code> = <code>ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG</code>, a 50 bp read</li>
<li><code>read_qscores</code> = <code>CCCBCGFGBGGGGGGGBGGGGGGGGG&gt;GGG1G=/GGGGGGGGGGGGGGGG</code></li>
</ul>
</section>
</section>
<section id="aln-.aln" class="level4">
<h4 class="anchored" data-anchor-id="aln-.aln">ALN (<code>.aln</code>)</h4>
<p>ALN files generated by ART Illumina consist of : - a header with the ART-Ilumina command used for the simulation (<code>@CM</code>) and info on each of the reference sequences used for the simulations (<code>@SQ</code>). Header always starts with <code>##ART_Illumina</code> and ends with <code>##Header End</code> : - the body with 3 lines for each read: 1. definition line with <code>readid</code>, - reference sequence identification number <code>refseqid</code>, - the position in the read in the reference sequence <code>aln_start_pos</code> - the strand the read was taken from <code>ref_seq_strand</code>. <code>+</code> for coding strand and <code>-</code> for template strand 2. aligned reference sequence, that is the sequence segment in the original reference corresponding to the read 3. aligned read sequence, that is the simmulated read sequence, where each bp corresponds to the reference sequence bp in the same position.</p>
<p>Example of a ALN file generated by ART Illumina (showing 5 reads):</p>
<pre class="ascii"><code>##ART_Illumina    read_length    50
@CM    /bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/cov_data/cov_virus_sequences_ten.fa -ss HS25 -l 50 -f 100 -o /home/vtec/projects/bio/metagentools/data/cov_simreads/single_10seq_50bp/single_10seq_50bp -rs 1674660835
@SQ    2591237:ncbi:1 1   MK211378    2591237    ncbi    1     Coronavirus BtRs-BetaCoV/YN2018D    30213
@SQ    11128:ncbi:2   2   LC494191    11128    ncbi    2     Bovine coronavirus    30942
@SQ    31631:ncbi:3   3   KY967361    31631    ncbi    3     Human coronavirus OC43        30661
@SQ    277944:ncbi:4  4   LC654455    277944    ncbi    4     Human coronavirus NL63    27516
@SQ    11120:ncbi:5   5   MN987231    11120    ncbi    5     Infectious bronchitis virus    27617
@SQ    28295:ncbi:6   6   KU893866    28295    ncbi    6     Porcine epidemic diarrhea virus    28043
@SQ    28295:ncbi:7   7   KJ645638    28295    ncbi    7     Porcine epidemic diarrhea virus    27998
@SQ    28295:ncbi:8   8   KJ645678    28295    ncbi    8     Porcine epidemic diarrhea virus    27998
@SQ    28295:ncbi:9   9   KR873434    28295    ncbi    9     Porcine epidemic diarrhea virus    28038
@SQ    1699095:ncbi:10 10  KT368904    1699095    ncbi    10     Camel alphacoronavirus    27395
##Header End
&gt;2591237:ncbi:1    2591237:ncbi:1-60400    14770    +
ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
ACAACTCCTATTCGTAGTTGAAGTTGTTGACAAATACTTTGATTGTTACG
&gt;2591237:ncbi:1    2591237:ncbi:1-60399    17012    -
GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC
GATCAATGTGGCATCTACAATACAGACAGCATGAAGCACCACCAAAGGAC
&gt;2591237:ncbi:1    2591237:ncbi:1-60398    9188    +
ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC
ATCTACCAGTGGTAGATGGGTTCTTAATAATGAACATTATAGAGCTCTAC
.....</code></pre>
</section>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
</section>
<section id="model-related-data" class="level2">
<h2 class="anchored" data-anchor-id="model-related-data">Model related data</h2>
<div id="cell-27" class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>pfs <span class="op">=</span> ProjectFileSystem()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-28" class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>pfs.readme(pfs.data <span class="op">/</span> <span class="st">'saved'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<hr>
</div>
<div class="cell-output cell-output-display cell-output-markdown">
<p>ReadMe file for directory <code>nbs-dev/data_dev/saved</code>:</p>
</div>
<div class="cell-output cell-output-display">
<hr>
</div>
<section id="saved-data-related-to-models" class="level3 cell-output cell-output-display cell-output-markdown">
<h3 class="anchored" data-anchor-id="saved-data-related-to-models">Saved data related to models</h3>
<p>This directory includes all data related to models and saved: - saved model parameters - saved datasets</p>
<p>For example: - <code>cnn_virus_original/pretrained_model.h5</code> is the saved model parameters for the CNN Virus model</p>
</section>
<div class="cell-output cell-output-display">
<hr>
</div>
</div>
</section>
</section>
<section id="parsing-sequence-files" class="level1">
<h1>Parsing sequence files</h1>
<p>The following classes make it easier to read and parse files of different formats into their underlying components to generated the training, validation, testing and inference datasets for the model.</p>
<p>Each class inherits from <a href="https://vtecftwy.github.io/metagentorch/core.html#textfilebasereader"><code>TextFileBaseReader</code></a> and adds:</p>
<ul>
<li>One or several text parsing method(s) to parse metadata according to a specific format</li>
<li>A file parsing method to extract metadata from all elements in the file, returning it as a key:value dictionary and optionally save the metadata as a json file.</li>
</ul>
<section id="fasta-file" class="level2">
<h2 class="anchored" data-anchor-id="fasta-file">FASTA file</h2>
<p>Extension of <a href="https://vtecftwy.github.io/metagentorch/core.html#textfilebasereader"><code>TextFileBaseReader</code></a> class for fasta sequence files.</p>
<p>Structure of a FASTA sequence file:</p>
<pre class="ascii"><code>&gt;definition line - format varies from dataset to dataset
sequence line: sequence of bases</code></pre>
<p>Example for the NCBI datasets:</p>
<pre class="ascii"><code>&gt;seqid accession taxonomyid source seqnb organism
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...
&gt;2591237:ncbi:1 MK211378    2591237 ncbi    1 Coronavirus BtRs-BetaCoV/YN2018D
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...</code></pre>
<div id="cell-33" class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>p2fasta <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/cov_virus_sequences_two.fa'</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> TextFileBaseReader(p2fasta, nlines<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, t <span class="kw">in</span> <span class="bu">enumerate</span>(fasta):</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>    txt <span class="op">=</span> t.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>)[:<span class="dv">80</span>] <span class="op">+</span> <span class="st">' ...'</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>txt<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D ...
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...
&gt;11128:ncbi:2   2   LC494191    11128   ncbi    Bovine coronavirus ...
CATCCCGCTTCACTGATCTCTTGTTAGATCTTTTCATAATCTAAACTTTATAAAAACATCCACTCCCTGTAGTCTATGCC ...</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L76" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="fastafilereader" class="level3">
<h3 class="anchored" data-anchor-id="fastafilereader">FastaFileReader</h3>
<blockquote class="blockquote">
<pre><code> FastaFileReader (path:str|pathlib.Path)</code></pre>
</blockquote>
<p><em>Wrap a FASTA file and retrieve its content in raw format and parsed format</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>path</td>
<td>str | pathlib.Path</td>
<td>path to the Fasta file</td>
</tr>
</tbody>
</table>
<p>As an iterator, <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#fastafilereader"><code>FastaFileReader</code></a> returns a <code>dict</code> at each step, as follows:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'definition line'</span>: <span class="st">'string in file as the definition line for the sequence'</span>,</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'sequence'</span>: <span class="st">'the full sequence'</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Illustration:</p>
<div id="cell-36" class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>p2fasta <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/cov_virus_sequences_two.fa'</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>iteration_output <span class="op">=</span> <span class="bu">next</span>(fasta)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(iteration_output[<span class="st">'definition line'</span>][:<span class="dv">80</span>], <span class="st">'...'</span>)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(iteration_output[<span class="st">'sequence'</span>][:<span class="dv">80</span>], <span class="st">'...'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D ...
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...</code></pre>
</div>
</div>
<div id="cell-37" class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"output type :     </span><span class="sc">{</span><span class="bu">type</span>(iteration_output)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"keys :            </span><span class="sc">{</span>iteration_output<span class="sc">.</span>keys()<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"definition line : </span><span class="sc">{</span>iteration_output[<span class="st">'definition line'</span>][:<span class="dv">80</span>]<span class="sc">}</span><span class="ss"> ...'"</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"sequence :       '</span><span class="sc">{</span>iteration_output[<span class="st">'sequence'</span>][:<span class="dv">100</span>]<span class="sc">}</span><span class="ss"> ...'"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>output type :     &lt;class 'dict'&gt;
keys :            dict_keys(['definition line', 'sequence'])
definition line : &gt;2591237:ncbi:1   1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D ...'
sequence :       'TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTAGCTGTCGCTCGGC ...'</code></pre>
</div>
</div>
<p>The <code>definition line</code> is a string, with tab separated values.</p>
<div id="cell-39" class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>display(iteration_output[<span class="st">'definition line'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'&gt;2591237:ncbi:1\t1\tMK211378\t2591237\tncbi\tCoronavirus BtRs-BetaCoV/YN2018D'</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L137" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="fastafilereader.review" class="level3">
<h3 class="anchored" data-anchor-id="fastafilereader.review">FastaFileReader.review</h3>
<blockquote class="blockquote">
<pre><code> FastaFileReader.review ()</code></pre>
</blockquote>
<p><em>Prints the first and last sequences and metadata in the fasta file and returns the nb or sequences</em></p>
<div id="cell-41" class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>nb_seqs <span class="op">=</span> fasta.review()</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>nb_seqs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>There are 2 sequences in this file

First Sequence:
&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...
{'seqid': '2591237:ncbi:1', 'taxonomyid': '2591237', 'source': 'ncbi', 'seqnb': '1', 'accession': 'MK211378', 'organism': 'Coronavirus BtRs-BetaCoV/YN2018D'}

Last Sequence:
&gt;11128:ncbi:2   2   LC494191    11128   ncbi    Bovine coronavirus
CATCCCGCTTCACTGATCTCTTGTTAGATCTTTTCATAATCTAAACTTTATAAAAACATCCACTCCCTGTAGTCTATGCC ...
{'seqid': '11128:ncbi:2', 'taxonomyid': '11128', 'source': 'ncbi', 'seqnb': '2', 'accession': 'LC494191', 'organism': 'Bovine coronavirus'}</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="32">
<pre><code>2</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L100" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="fastafilereader.print_first_chunks" class="level3">
<h3 class="anchored" data-anchor-id="fastafilereader.print_first_chunks">FastaFileReader.print_first_chunks</h3>
<blockquote class="blockquote">
<pre><code> FastaFileReader.print_first_chunks (nchunks:int=3)</code></pre>
</blockquote>
<p><em>Print the first <code>nchunks</code> chunks of text from the file</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nchunks</td>
<td>int</td>
<td>3</td>
<td>number of chunks to print out</td>
</tr>
</tbody>
</table>
<p>This is convenient to quickly discover and explore new fasta files in raw text format:</p>
<div id="cell-44" class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>fasta.print_first_chunks(nchunks<span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Sequence 1:
&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D
TATTAGGTTTTCTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAAT ...

Sequence 2:
&gt;11128:ncbi:2   2   LC494191    11128   ncbi    Bovine coronavirus
CATCCCGCTTCACTGATCTCTTGTTAGATCTTTTCATAATCTAAACTTTATAAAAACATCCACTCCCTGTAGTCTATGCC ...</code></pre>
</div>
</div>
</section>
<section id="parsing-metadata" class="level3">
<h3 class="anchored" data-anchor-id="parsing-metadata">Parsing metadata</h3>
<p>The class also provides methods to parse metadata from the file content (definition line, headers, …).</p>
<p>A regex pattern is used for parsing metadata fom the definition lines in the reference sequence fasta file.</p>
<p>Below, we parse the data from the definition line of our Corona virus NCBI dataset (rule <code>fasta_ncbi_std</code>):</p>
<p>Sequence 1:</p>
<ul>
<li>Definition Line:</li>
</ul>
<pre class="ascii"><code>&gt;2591237:ncbi:1 [MK211378]  2591237 ncbi    1 [MK211378] 2591237    Coronavirus YN2018D     scientific name</code></pre>
<ul>
<li>Metadata:
<ul>
<li><code>seqid</code> = <code>2591237:ncbi:1</code></li>
<li><code>taxonomyid</code> = <code>2591237</code></li>
<li><code>source</code> = <code>ncbi</code></li>
<li><code>seqnb</code> = <code>1</code></li>
<li><code>accession</code> = <code>MK211378</code></li>
<li><code>species</code> = <code>Coronavirus BtRs-BetaCoV/YN2018D</code></li>
</ul></li>
</ul>
<p>Sequence 2:</p>
<ul>
<li>Definition Line</li>
</ul>
<pre class="ascii"><code>    &gt;11128:ncbi:2 [LC494191]</code></pre>
<ul>
<li>Metadata:
<ul>
<li><code>seqid</code> = <code>11128:ncbi:2</code></li>
<li><code>taxonomyid</code> = <code>11128</code></li>
<li><code>source</code> = <code>ncbi</code></li>
<li><code>seqnb</code> = <code>2</code></li>
<li><code>accession</code> = <code>LC494191</code></li>
<li><code>species</code> = <code>''</code></li>
</ul></li>
</ul>
<p><a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#fastafilereader"><code>FastaFileReader</code></a> offers: - <code>parse_text</code> a method to parse the metadata - an option to set a default “parsing rule” for one instance with <code>set_parsing_rules</code>. - <code>parse_file</code> a method to parse the metadata from all sequences in the file and save it as a json file.</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/core.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="textfilebasereader.parse_text" class="level3">
<h3 class="anchored" data-anchor-id="textfilebasereader.parse_text">TextFileBaseReader.parse_text</h3>
<blockquote class="blockquote">
<pre><code> TextFileBaseReader.parse_text (txt:str, pattern:str|None=None)</code></pre>
</blockquote>
<p><em>Parse text using regex pattern with groups. Return a metadata dictionary.</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>txt</td>
<td>str</td>
<td></td>
<td>text to parse</td>
</tr>
<tr class="even">
<td>pattern</td>
<td>str | None</td>
<td>None</td>
<td>If None, uses standard regex pattern to extract metadata, otherwise, uses passed regex</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td><strong>parsed metadata in key/value format</strong></td>
</tr>
</tbody>
</table>
<p>Running the parser function with specifically defined <code>pattern</code> and <code>keys</code>.</p>
<div id="cell-52" class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>dfn_line, sequence <span class="op">=</span> <span class="bu">next</span>(fasta).values()</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(dfn_line.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D</code></pre>
</div>
</div>
<div id="cell-53" class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>pattern <span class="op">=</span> <span class="vs">r"^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*(?P=seqnb)[\s\t](?P&lt;accession&gt;[\w\d]*)([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t][\s\t]*(?P&lt;organism&gt;[\w\s\-\_\/]*))?"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-54" class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>fasta.parse_text(dfn_line, pattern<span class="op">=</span>pattern)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="39">
<pre><code>{'seqid': '2591237:ncbi:1',
 'taxonomyid': '2591237',
 'source': 'ncbi',
 'seqnb': '1',
 'accession': 'MK211378',
 'organism': 'Coronavirus BtRs-BetaCoV/YN2018D'}</code></pre>
</div>
</div>
<p>When a <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#fastafilereader"><code>FastaFileReader</code></a> instance is created, all existing rules in the file <code>default_parsing_rules.json</code> are tested on the first definition line of the fasta file and the one rule that parses the most matches will be selected automatically and saved in instance attributes <code>re_rule_name</code>, <code>re_pattern</code> and <code>re_keys</code>.</p>
<p><code>parse_file</code> extract metadata from each definition line in the fasta file and return a dictionary with all metadata.</p>
<div id="cell-56" class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta.re_rule_name)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta.re_pattern)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta.re_keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fasta_ncbi_std
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*(?P=seqnb)[\s\t](?P&lt;accession&gt;[\w\d]*)([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t][\s\t]*(?P&lt;organism&gt;[\w\s\-\_/]*))?
['seqid', 'taxonomyid', 'source', 'seqnb', 'accession', 'organism']</code></pre>
</div>
</div>
<div id="cell-57" class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>fasta.parse_text(dfn_line)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="41">
<pre><code>{'seqid': '2591237:ncbi:1',
 'taxonomyid': '2591237',
 'source': 'ncbi',
 'seqnb': '1',
 'accession': 'MK211378',
 'organism': 'Coronavirus BtRs-BetaCoV/YN2018D'}</code></pre>
</div>
</div>
<p>When another fasta file, which has another definition line structure, is used, another parsing rule is selected.</p>
<div id="cell-59" class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>p2other <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/another_sequence.fa'</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> p2other.is_file()</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>it2 <span class="op">=</span> FastaFileReader(path<span class="op">=</span>p2other)</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>dfn_line, sequence <span class="op">=</span> <span class="bu">next</span>(it2).values()</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(dfn_line.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;1 dna_rm:primary_assembly primary_assembly:mRhiFer1_v1.p:1:1:124933378:1 REF</code></pre>
</div>
</div>
<div id="cell-60" class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(it2.re_rule_name)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(it2.re_pattern)</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(it2.re_keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fasta_rhinolophus_ferrumequinum
^&gt;\d[\s\t](?P&lt;seq_type&gt;dna_rm):(?P&lt;id_type&gt;[\w\_]*)[\s\w](?P=id_type):(?P&lt;assy&gt;[\w\d\_]*)\.(?P&lt;seq_level&gt;[\w]*):\d*:\d*:(?P&lt;taxonomy&gt;\d*):(?P&lt;id&gt;\d*)[\s    ]REF$
['seq_type', 'id_type', 'assy', 'seq_level', 'taxonomy', 'id']</code></pre>
</div>
</div>
<div id="cell-61" class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>pprint(it2.parse_text(dfn_line))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'assy': 'mRhiFer1_v1',
 'id': '1',
 'id_type': 'primary_assembly',
 'seq_level': 'p',
 'seq_type': 'dna_rm',
 'taxonomy': '124933378'}</code></pre>
</div>
</div>
<p>This rule selection is performed by the class method <code>set_parsing_rule</code>. The method can also be called with specific <code>pattern</code> and <code>keys</code> to force parsing rule not yet saved in the json file.</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/core.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="textfilebasereader.set_parsing_rules" class="level3">
<h3 class="anchored" data-anchor-id="textfilebasereader.set_parsing_rules">TextFileBaseReader.set_parsing_rules</h3>
<blockquote class="blockquote">
<pre><code> TextFileBaseReader.set_parsing_rules (pattern:str|None=None,
                                       verbose:bool=False)</code></pre>
</blockquote>
<p>*Set the standard regex parsing rule for the file.</p>
<p>Rules can be set:</p>
<ol type="1">
<li>manually by passing specific custom values for <code>pattern</code> and <code>keys</code></li>
<li>automatically, by testing all parsing rules saved in <code>parsing_rule.json</code></li>
</ol>
<p>Automatic selection of parsing rules works by testing each rule saved in <code>parsing_rule.json</code> on the first definition line of the fasta file, and selecting the one rule that generates the most metadata matches.</p>
<p>Rules consists of two parameters:</p>
<ul>
<li>The regex pattern including one <code>group</code> for each metadata item, e.g <code>(?P&lt;group_name&gt;regex_code)</code></li>
<li>The list of keys, i.e.&nbsp;the list with the name of each regex groups, used as key in the metadata dictionary</li>
</ul>
<p>This method updates the three following class attributes: <code>re_rule_name</code>, <code>re_pattern</code>, <code>re_keys</code>*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pattern</td>
<td>str | None</td>
<td>None</td>
<td>regex pattern to apply to parse the text, search in parsing rules json if None</td>
</tr>
<tr class="even">
<td>verbose</td>
<td>bool</td>
<td>False</td>
<td>when True, provides information on each rule</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-64" class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>dfn_line, sequence <span class="op">=</span> <span class="bu">next</span>(fasta).values()</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"definition line: '</span><span class="sc">{</span>dfn_line[:<span class="op">-</span><span class="dv">1</span>]<span class="sc">}</span><span class="ss">'"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>definition line: '&gt;2591237:ncbi:1   1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018'</code></pre>
</div>
</div>
<p>Automatic parsing works by testing each saved rule for the value of <code>definition line</code> in the first sequence in the fasta file.</p>
<div id="cell-66" class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"key for text to parse: </span><span class="sc">{</span>fasta<span class="sc">.</span>text_to_parse_key<span class="sc">}</span><span class="ch">\n</span><span class="ss">"</span>)</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>fasta.reset_iterator()</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Text to parse for testing (extracted from first iteration):'</span>)</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">next</span>(fasta)[fasta.text_to_parse_key])</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>()</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>fasta.set_parsing_rules(verbose<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>key for text to parse: definition line

Text to parse for testing (extracted from first iteration):
&gt;2591237:ncbi:1 1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D

--------------------------------------------------------------------------------
6 matches generated with rule &lt;fasta_ncbi_std&gt; 
--------------------------------------------------------------------------------
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*(?P=seqnb)[\s\t](?P&lt;accession&gt;[\w\d]*)([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t][\s\t]*(?P&lt;organism&gt;[\w\s\-\_/]*))?
['seqid', 'taxonomyid', 'source', 'seqnb', 'accession', 'organism']
{'seqid': '2591237:ncbi:1', 'taxonomyid': '2591237', 'source': 'ncbi', 'seqnb': '1', 'accession': 'MK211378', 'organism': 'Coronavirus BtRs-BetaCoV/YN2018D'}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;fastq_art_illumina_ncbi_std&gt; 
--------------------------------------------------------------------------------
^@(?P&lt;readid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*)-(?P&lt;readnb&gt;\d*(\/\d)?))$
['readid', 'reftaxonomyid', 'refsource', 'refseqnb', 'readnb']
{'readid': '', 'reftaxonomyid': '', 'refsource': '', 'refseqnb': '', 'readnb': ''}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;aln_art_illumina_ncbi_std&gt; 
--------------------------------------------------------------------------------
^&gt;(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))(\s| )*(?P&lt;readid&gt;(?P=reftaxonomyid):(?P=refsource):(?P=refseqnb)-(?P&lt;readnb&gt;\d*(\/\d(-\d)?)?))(\s|\t)(?P&lt;aln_start_pos&gt;\d*)(\s|\t)(?P&lt;refseq_strand&gt;(-|\+))$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'readid', 'readnb', 'aln_start_pos', 'refseq_strand']
{'refseqid': '', 'reftaxonomyid': '', 'refsource': '', 'refseqnb': '', 'readid': '', 'readnb': '', 'aln_start_pos': '', 'refseq_strand': ''}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;aln_art_illumina-refseq-ncbi-std&gt; 
--------------------------------------------------------------------------------
^@SQ[\t\s]*(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))[\t\s]*(?P=refseqnb)[\t\s]*(?P&lt;refseq_accession&gt;[\w\d]*)[\t\s]*(?P=reftaxonomyid)[\t\s]*(?P=refsource)[\t\s](?P&lt;organism&gt;.*)[\t\s](?P&lt;refseq_length&gt;\d*)$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'refseq_accession', 'organism', 'refseq_length']
{'refseqid': '', 'reftaxonomyid': '', 'refsource': '', 'refseqnb': '', 'refseq_accession': '', 'organism': '', 'refseq_length': ''}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;fasta_ncbi_cov&gt; 
--------------------------------------------------------------------------------
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*\[(?P&lt;accession&gt;[\w\d]*)\]([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t]*(?P=seqnb)[\s\t]*\[(?P=accession)\][\s\t]*(?P=taxonomyid)[\s\t]*(?P&lt;organism&gt;[\w\s\-\_\/]*))?
['seqid', 'taxonomyid', 'source', 'seqnb', 'accession', 'organism']
{'seqid': '', 'taxonomyid': '', 'source': '', 'seqnb': '', 'accession': '', 'organism': ''}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;fasta_rhinolophus_ferrumequinum&gt; 
--------------------------------------------------------------------------------
^&gt;\d[\s\t](?P&lt;seq_type&gt;dna_rm):(?P&lt;id_type&gt;[\w\_]*)[\s\w](?P=id_type):(?P&lt;assy&gt;[\w\d\_]*)\.(?P&lt;seq_level&gt;[\w]*):\d*:\d*:(?P&lt;taxonomy&gt;\d*):(?P&lt;id&gt;\d*)[\s    ]REF$
['seq_type', 'id_type', 'assy', 'seq_level', 'taxonomy', 'id']
{'seq_type': '', 'id_type': '', 'assy': '', 'seq_level': '', 'taxonomy': '', 'id': ''}
--------------------------------------------------------------------------------
Selected rule with most matches: fasta_ncbi_std</code></pre>
</div>
</div>
<p>If no saved rule generates a match, <code>re_rule_name</code>, <code>re_pattern</code> and <code>re_keys</code> remain <code>None</code> and a warning message is issued to ask user to add a parsing rule manually.</p>
<div id="cell-68" class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>p2nomatch <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/sequences_two_no_matching_rule.fa'</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>fasta2 <span class="op">=</span> FastaFileReader(p2nomatch)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/vtec/projects/bio/metagentorch/metagentorch/core.py:651: UserWarning: 
        None of the saved parsing rules were able to extract metadata from the first line in this file.
        You must set a custom rule (pattern + keys) before parsing text, by using:
            `self.set_parsing_rules(custom_pattern)`
                
  warnings.warn(msg, category=UserWarning)</code></pre>
</div>
</div>
<div id="cell-69" class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>fasta2.re_rule_name <span class="kw">is</span> <span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="49">
<pre><code>True</code></pre>
</div>
</div>
<p>But we still can set a standard rule manually, by passing a re pattern and the corresponding list of keys.</p>
<div id="cell-71" class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>pat <span class="op">=</span> <span class="vs">r"^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))\s*(?P&lt;text&gt;[\w\s]*)$"</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>keys <span class="op">=</span> <span class="st">"seqid taxonomyid source seqnb text"</span>.split()</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>fasta2.set_parsing_rules(pattern<span class="op">=</span>pat)</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta2.re_rule_name)</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta2.re_pattern)</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fasta2.re_keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Custom Rule
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))\s*(?P&lt;text&gt;[\w\s]*)$
['seqid', 'taxonomyid', 'source', 'seqnb', 'text']</code></pre>
</div>
</div>
<div id="cell-72" class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>fasta2.reset_iterator()</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>dfn_line, sequence <span class="op">=</span> <span class="bu">next</span>(fasta2).values()</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"definition line: '</span><span class="sc">{</span>dfn_line[:<span class="op">-</span><span class="dv">1</span>]<span class="sc">}</span><span class="ss">'"</span>)</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>fasta2.parse_text(dfn_line)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>definition line: '&gt;2591237:ncbi:1 this sequence does not match any saved parsing rul'</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="51">
<pre><code>{'seqid': '2591237:ncbi:1',
 'taxonomyid': '2591237',
 'source': 'ncbi',
 'seqnb': '1',
 'text': 'this sequence does not match any saved parsing rule'}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L113" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="fastafilereader.parse_file" class="level3">
<h3 class="anchored" data-anchor-id="fastafilereader.parse_file">FastaFileReader.parse_file</h3>
<blockquote class="blockquote">
<pre><code> FastaFileReader.parse_file (add_seq:bool=False, save_json:bool=False)</code></pre>
</blockquote>
<p><em>Read fasta file and return a dictionary with definition line metadata and optionally sequences</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>add_seq</td>
<td>bool</td>
<td>False</td>
<td>When True, add the full sequence to the parsed metadata dictionary</td>
</tr>
<tr class="even">
<td>save_json</td>
<td>bool</td>
<td>False</td>
<td>When True, save the file metadata as a json file of same stem name</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td><strong>Metadata as Key/Values pairs</strong></td>
</tr>
</tbody>
</table>
<div id="cell-74" class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>pprint(fasta.parse_file())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'11128:ncbi:2': {'accession': 'LC494191',
                  'organism': 'Bovine coronavirus',
                  'seqid': '11128:ncbi:2',
                  'seqnb': '2',
                  'source': 'ncbi',
                  'taxonomyid': '11128'},
 '2591237:ncbi:1': {'accession': 'MK211378',
                    'organism': 'Coronavirus BtRs-BetaCoV/YN2018D',
                    'seqid': '2591237:ncbi:1',
                    'seqnb': '1',
                    'source': 'ncbi',
                    'taxonomyid': '2591237'}}</code></pre>
</div>
</div>
<div id="cell-75" class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>fasta.parse_file(save_json<span class="op">=</span><span class="va">True</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Metadata for 'cov_virus_sequences_two.fa'&gt; saved as &lt;cov_virus_sequences_two_metadata.json&gt; in  
/home/vtec/projects/bio/metagentorch/nbs-dev/data_dev/ncbi/refsequences/cov
</code></pre>
</div>
</div>
<div id="cell-76" class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'../default_parsing_rules.json'</span>, <span class="st">'r'</span>) <span class="im">as</span> fp:</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>    pprint(json.load(fp), width<span class="op">=</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'aln_art_illumina-refseq-ncbi-std': {'keys': 'refseqid '
                                              'reftaxonomyid '
                                              'refsource '
                                              'refseqnb '
                                              'refseq_accession '
                                              'organism '
                                              'refseq_length',
                                      'pattern': '^@SQ[\\t\\s]*(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\\d*):(?P&lt;refsource&gt;\\w*):(?P&lt;refseqnb&gt;\\d*))[\\t\\s]*(?P=refseqnb)[\\t\\s]*(?P&lt;refseq_accession&gt;[\\w\\d]*)[\\t\\s]*(?P=reftaxonomyid)[\\t\\s]*(?P=refsource)[\\t\\s](?P&lt;organism&gt;.*)[\\t\\s](?P&lt;refseq_length&gt;\\d*)$'},
 'aln_art_illumina_ncbi_std': {'keys': 'refseqid '
                                       'reftaxonomyid '
                                       'refsource '
                                       'refseqnb '
                                       'readid '
                                       'readnb '
                                       'aln_start_pos '
                                       'refseq_strand',
                               'pattern': '^&gt;(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\\d*):(?P&lt;refsource&gt;\\w*):(?P&lt;refseqnb&gt;\\d*))(\\s|\t'
                                          ')*(?P&lt;readid&gt;(?P=reftaxonomyid):(?P=refsource):(?P=refseqnb)-(?P&lt;readnb&gt;\\d*(\\/\\d(-\\d)?)?))(\\s|\\t)(?P&lt;aln_start_pos&gt;\\d*)(\\s|\\t)(?P&lt;refseq_strand&gt;(-|\\+))$'},
 'fasta_ncbi_cov': {'keys': 'seqid '
                            'taxonomyid '
                            'source '
                            'accession '
                            'seqnb '
                            'organism',
                    'pattern': '^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\\d*))[\\s\\t]*\\[(?P&lt;accession&gt;[\\w\\d]*)\\]([\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P=source)[\\s\\t]*(?P=seqnb)[\\s\\t]*\\[(?P=accession)\\][\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P&lt;organism&gt;[\\w\\s\\-\\_\\/]*))?'},
 'fasta_ncbi_std': {'keys': 'seqid '
                            'taxonomyid '
                            'source '
                            'accession '
                            'seqnb '
                            'organism',
                    'pattern': '^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\\d*))[\\s\\t]*(?P=seqnb)[\\s\\t](?P&lt;accession&gt;[\\w\\d]*)([\\s\\t]*(?P=taxonomyid)[\\s\\t]*(?P=source)[\\s\\t][\\s\\t]*(?P&lt;organism&gt;[\\w\\s\\-\\_/]*))?'},
 'fasta_rhinolophus_ferrumequinum': {'keys': 'seq_type '
                                             'id_type '
                                             'assy '
                                             'seq_level '
                                             'taxonomy '
                                             'id',
                                     'pattern': '^&gt;\\d[\\s\\t](?P&lt;seq_type&gt;dna_rm):(?P&lt;id_type&gt;[\\w\\_]*)[\\s\\w](?P=id_type):(?P&lt;assy&gt;[\\w\\d\\_]*)\\.(?P&lt;seq_level&gt;[\\w]*):\\d*:\\d*:(?P&lt;taxonomy&gt;\\d*):(?P&lt;id&gt;\\d*)[\\s\t'
                                                ']REF$'},
 'fastq_art_illumina_ncbi_std': {'keys': 'readid '
                                         'reftaxonomyid '
                                         'refsource '
                                         'refseqnb '
                                         'readnb',
                                 'pattern': '^@(?P&lt;readid&gt;(?P&lt;reftaxonomyid&gt;\\d*):(?P&lt;refsource&gt;\\w*):(?P&lt;refseqnb&gt;\\d*)-(?P&lt;readnb&gt;\\d*(\\/\\d)?))$'}}</code></pre>
</div>
</div>
<div id="cell-77" class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>p2fasta <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/refsequences/cov/cov_virus_sequence_one.fa'</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>fasta <span class="op">=</span> FastaFileReader(p2fasta)</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>fasta_meta <span class="op">=</span> fasta.parse_file(save_json<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>pprint(fasta_meta)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Metadata for 'cov_virus_sequence_one.fa'&gt; saved as &lt;cov_virus_sequence_one_metadata.json&gt; in  
/home/vtec/projects/bio/metagentorch/nbs-dev/data_dev/ncbi/refsequences/cov

{'2591237:ncbi:1': {'accession': 'MK211378',
                    'organism': 'Coronavirus BtRs-BetaCoV/YN2018D',
                    'seqid': '2591237:ncbi:1',
                    'seqnb': '1',
                    'source': 'ncbi',
                    'taxonomyid': '2591237'}}</code></pre>
</div>
</div>
</section>
</section>
<section id="fastq-file" class="level2">
<h2 class="anchored" data-anchor-id="fastq-file">FASTQ file</h2>
<p>Extension of <a href="https://vtecftwy.github.io/metagentorch/core.html#textfilebasereader"><code>TextFileBaseReader</code></a> class for fastq sequence files.</p>
<p>Structure of a FASTQ sequence file:</p>
<div id="cell-80" class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>p2fastq <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads/cov/single_1seq_150bp/single_1seq_150bp.fq'</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>fastq <span class="op">=</span> TextFileBaseReader(p2fastq, nlines<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, t <span class="kw">in</span> <span class="bu">enumerate</span>(fastq):</span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>    txt <span class="op">=</span> t.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>)[:<span class="dv">80</span>]</span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>txt<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb87-7"><a href="#cb87-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">11</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>@2591237:ncbi:1-40200
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC
+
CCCGGGCGGGGGCJGJJJGJJGJJJGJGGJGJJJGJGGGGGGGGCJGJGGGGGJJJJGCCGGGGGJCGCGJGJCG=GGGG
@2591237:ncbi:1-40199
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGA
+
=CCGGGGCGGGGGJJGJJGJGJG=GJJGJCGJJJCJ=JJJJGGJJCJGJGG=JGC1JJGG8GCJCGGGCGG(GCGGCGC=
@2591237:ncbi:1-40198
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT
+
C=CGGGGGGGGGGCJJJJ=JJJJJJJJJJJGGJJJJ1GJJ8GJJGGGJGGJJC=JJGGGCCGG88GG=GGGGGGCJGGGG</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L160" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="fastqfilereader" class="level3">
<h3 class="anchored" data-anchor-id="fastqfilereader">FastqFileReader</h3>
<blockquote class="blockquote">
<pre><code> FastqFileReader (path:str|pathlib.Path)</code></pre>
</blockquote>
<p><em>Iterator going through a fastq file’s sequences and return each section + prob error as a dict</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>path</td>
<td>str | pathlib.Path</td>
<td>path to the fastq file</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>key/value with keys: definition line; sequence; q score; prob error</strong></td>
</tr>
</tbody>
</table>
<div id="cell-82" class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>fastq <span class="op">=</span> FastqFileReader(p2fastq)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>iteration_output <span class="op">=</span> <span class="bu">next</span>(fastq)</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">type</span>(iteration_output))</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(iteration_output.keys())</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Definition line:  </span><span class="sc">{</span>iteration_output[<span class="st">'definition line'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Read sequence:    </span><span class="sc">{</span>iteration_output[<span class="st">'sequence'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Q scores (ASCII): </span><span class="sc">{</span>iteration_output[<span class="st">'read_qscores'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Prob error:       </span><span class="sc">{</span><span class="st">','</span><span class="sc">.</span>join([<span class="ss">f'</span><span class="sc">{</span>p<span class="sc">:.4f}</span><span class="ss">'</span> <span class="cf">for</span> p <span class="kw">in</span> iteration_output[<span class="st">'probs error'</span>]])<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'dict'&gt;
dict_keys(['definition line', 'sequence', 'read_qscores', 'probs error'])
Definition line:  @2591237:ncbi:1-40200
Read sequence:    TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAGTTTCAAGGAACTTTTAGTGTATGCTGCTGATCCAGCCATGCATGCAGCTT
Q scores (ASCII): CCCGGGCGGGGGCJGJJJGJJGJJJGJGGJGJJJGJGGGGGGGGCJGJGGGGGJJJJGCCGGGGGJCGCGJGJCG=GGGG=CGGGGGG1GCGCGGGGCCGJC8GGGGGGGGGGGCGGGGGGGGGGGC8GGGGGGCGGC1GGGCGGGGGCC
Prob error:       0.0004,0.0004,0.0004,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0001,0.0002,0.0001,0.0001,0.0001,0.0002,0.0001,0.0001,0.0002,0.0001,0.0001,0.0001,0.0002,0.0001,0.0002,0.0002,0.0001,0.0002,0.0001,0.0001,0.0001,0.0002,0.0001,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0001,0.0002,0.0001,0.0002,0.0002,0.0002,0.0002,0.0002,0.0001,0.0001,0.0001,0.0001,0.0002,0.0004,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0001,0.0004,0.0002,0.0004,0.0002,0.0001,0.0002,0.0001,0.0004,0.0002,0.0016,0.0002,0.0002,0.0002,0.0002,0.0016,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0251,0.0002,0.0004,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0004,0.0004,0.0002,0.0001,0.0004,0.0050,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0050,0.0002,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0004,0.0251,0.0002,0.0002,0.0002,0.0004,0.0002,0.0002,0.0002,0.0002,0.0002,0.0004,0.0004</code></pre>
</div>
</div>
<p>Five largest probabilities of error:</p>
<div id="cell-84" class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>np.sort(iteration_output[<span class="st">'probs error'</span>])[<span class="op">-</span><span class="dv">5</span>:]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="73">
<pre><code>array([0.00158489, 0.00501187, 0.00501187, 0.02511886, 0.02511886])</code></pre>
</div>
</div>
<div id="cell-85" class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>np.argsort(iteration_output[<span class="st">'probs error'</span>])[<span class="op">-</span><span class="dv">5</span>:]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="74">
<pre><code>array([ 80, 127, 102, 138,  88])</code></pre>
</div>
</div>
<div id="cell-86" class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>dfn_line <span class="op">=</span> iteration_output[<span class="st">'definition line'</span>]</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>meta <span class="op">=</span> fastq.parse_text(dfn_line)</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>meta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="75">
<pre><code>{'readid': '2591237:ncbi:1-40200',
 'reftaxonomyid': '2591237',
 'refsource': 'ncbi',
 'refseqnb': '1',
 'readnb': '40200'}</code></pre>
</div>
</div>
<div id="cell-87" class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>fastq <span class="op">=</span> FastqFileReader(p2fastq)</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="bu">next</span>(fastq).keys()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="76">
<pre><code>dict_keys(['definition line', 'sequence', 'read_qscores', 'probs error'])</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L201" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="fastqfilereader.parse_file" class="level3">
<h3 class="anchored" data-anchor-id="fastqfilereader.parse_file">FastqFileReader.parse_file</h3>
<blockquote class="blockquote">
<pre><code> FastqFileReader.parse_file (add_readseq:bool=False,
                             add_qscores:bool=False,
                             add_probs_error:bool=False,
                             save_json:bool=False)</code></pre>
</blockquote>
<p><em>Read fastq file, return a dict with definition line metadata and optionally read sequence and q scores, …</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>add_readseq</td>
<td>bool</td>
<td>False</td>
<td>When True, add the full sequence to the parsed metadata dictionary</td>
</tr>
<tr class="even">
<td>add_qscores</td>
<td>bool</td>
<td>False</td>
<td>Add the read ASCII Q Scores to the parsed dictionary when True</td>
</tr>
<tr class="odd">
<td>add_probs_error</td>
<td>bool</td>
<td>False</td>
<td>Add the read probability of error to the parsed dictionary when True</td>
</tr>
<tr class="even">
<td>save_json</td>
<td>bool</td>
<td>False</td>
<td>When True, save the file metadata as a json file of same stem name</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td><strong>Metadata as Key/Values pairs</strong></td>
</tr>
</tbody>
</table>
<div id="cell-89" class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>parsed <span class="op">=</span> fastq.parse_file(add_readseq<span class="op">=</span><span class="va">False</span>, add_qscores<span class="op">=</span><span class="va">False</span>, add_probs_error<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (k, v) <span class="kw">in</span> <span class="bu">enumerate</span>(parsed.items()):</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(k)</span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>    pprint(v)</span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;=</span><span class="dv">3</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>2591237:ncbi:1-40200
{'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40199
{'readid': '2591237:ncbi:1-40199',
 'readnb': '40199',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40198
{'readid': '2591237:ncbi:1-40198',
 'readnb': '40198',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40197
{'readid': '2591237:ncbi:1-40197',
 'readnb': '40197',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}</code></pre>
</div>
</div>
<div id="cell-90" class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> fastq.parse_file(add_readseq<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(metadata).T</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>df.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="79">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">readid</th>
<th data-quarto-table-cell-role="th">reftaxonomyid</th>
<th data-quarto-table-cell-role="th">refsource</th>
<th data-quarto-table-cell-role="th">refseqnb</th>
<th data-quarto-table-cell-role="th">readnb</th>
<th data-quarto-table-cell-role="th">readseq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40200</td>
<td>2591237:ncbi:1-40200</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40200</td>
<td>TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCG...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40199</td>
<td>2591237:ncbi:1-40199</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40199</td>
<td>TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40198</td>
<td>2591237:ncbi:1-40198</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40198</td>
<td>TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTC...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40197</td>
<td>2591237:ncbi:1-40197</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40197</td>
<td>TAATCACTGATAGCAGCATTGCCATCCTGAGCAAAGAAGAAGTGTT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40196</td>
<td>2591237:ncbi:1-40196</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40196</td>
<td>CTAATGTCAGTACGCCTACAATGCCTGCATCACGCATAGCATCGCA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40195</td>
<td>2591237:ncbi:1-40195</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40195</td>
<td>AAGCTGAAGCATACATAACACAGTCCTTAAGCCGATAACCAGACAA...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40194</td>
<td>2591237:ncbi:1-40194</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40194</td>
<td>AGTGGAAGAACTTCACCGTCAAGATGAAACTCGACGGGGCTCTCCA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40193</td>
<td>2591237:ncbi:1-40193</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40193</td>
<td>GCGTCTCGAGTGCTTCGAGTTCACCGTTCTTGAGAACAACCTCCTC...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40192</td>
<td>2591237:ncbi:1-40192</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40192</td>
<td>CTGGTAGTATCTAAGGCTCCACTGAAATACTTGTACTTGTTATATA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2591237:ncbi:1-40191</td>
<td>2591237:ncbi:1-40191</td>
<td>2591237</td>
<td>ncbi</td>
<td>1</td>
<td>40191</td>
<td>GTCTCTATCTGTAGTACTATGACAAATAGACAGTTTCATCAGAAAT...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-91" class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>fastq.set_parsing_rules(verbose<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--------------------------------------------------------------------------------
0 matches generated with rule &lt;fasta_ncbi_std&gt; 
--------------------------------------------------------------------------------
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*(?P=seqnb)[\s\t](?P&lt;accession&gt;[\w\d]*)([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t][\s\t]*(?P&lt;organism&gt;[\w\s\-\_/]*))?
['seqid', 'taxonomyid', 'source', 'seqnb', 'accession', 'organism']
{'seqid': '', 'taxonomyid': '', 'source': '', 'seqnb': '', 'accession': '', 'organism': ''}
--------------------------------------------------------------------------------
5 matches generated with rule &lt;fastq_art_illumina_ncbi_std&gt; 
--------------------------------------------------------------------------------
^@(?P&lt;readid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*)-(?P&lt;readnb&gt;\d*(\/\d)?))$
['readid', 'reftaxonomyid', 'refsource', 'refseqnb', 'readnb']
{'readid': '2591237:ncbi:1-40200', 'reftaxonomyid': '2591237', 'refsource': 'ncbi', 'refseqnb': '1', 'readnb': '40200'}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;aln_art_illumina_ncbi_std&gt; 
--------------------------------------------------------------------------------
^&gt;(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))(\s| )*(?P&lt;readid&gt;(?P=reftaxonomyid):(?P=refsource):(?P=refseqnb)-(?P&lt;readnb&gt;\d*(\/\d(-\d)?)?))(\s|\t)(?P&lt;aln_start_pos&gt;\d*)(\s|\t)(?P&lt;refseq_strand&gt;(-|\+))$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'readid', 'readnb', 'aln_start_pos', 'refseq_strand']
{'refseqid': '', 'reftaxonomyid': '', 'refsource': '', 'refseqnb': '', 'readid': '', 'readnb': '', 'aln_start_pos': '', 'refseq_strand': ''}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;aln_art_illumina-refseq-ncbi-std&gt; 
--------------------------------------------------------------------------------
^@SQ[\t\s]*(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))[\t\s]*(?P=refseqnb)[\t\s]*(?P&lt;refseq_accession&gt;[\w\d]*)[\t\s]*(?P=reftaxonomyid)[\t\s]*(?P=refsource)[\t\s](?P&lt;organism&gt;.*)[\t\s](?P&lt;refseq_length&gt;\d*)$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'refseq_accession', 'organism', 'refseq_length']
{'refseqid': '', 'reftaxonomyid': '', 'refsource': '', 'refseqnb': '', 'refseq_accession': '', 'organism': '', 'refseq_length': ''}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;fasta_ncbi_cov&gt; 
--------------------------------------------------------------------------------
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*\[(?P&lt;accession&gt;[\w\d]*)\]([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t]*(?P=seqnb)[\s\t]*\[(?P=accession)\][\s\t]*(?P=taxonomyid)[\s\t]*(?P&lt;organism&gt;[\w\s\-\_\/]*))?
['seqid', 'taxonomyid', 'source', 'seqnb', 'accession', 'organism']
{'seqid': '', 'taxonomyid': '', 'source': '', 'seqnb': '', 'accession': '', 'organism': ''}
--------------------------------------------------------------------------------
0 matches generated with rule &lt;fasta_rhinolophus_ferrumequinum&gt; 
--------------------------------------------------------------------------------
^&gt;\d[\s\t](?P&lt;seq_type&gt;dna_rm):(?P&lt;id_type&gt;[\w\_]*)[\s\w](?P=id_type):(?P&lt;assy&gt;[\w\d\_]*)\.(?P&lt;seq_level&gt;[\w]*):\d*:\d*:(?P&lt;taxonomy&gt;\d*):(?P&lt;id&gt;\d*)[\s    ]REF$
['seq_type', 'id_type', 'assy', 'seq_level', 'taxonomy', 'id']
{'seq_type': '', 'id_type': '', 'assy': '', 'seq_level': '', 'taxonomy': '', 'id': ''}
--------------------------------------------------------------------------------
Selected rule with most matches: fastq_art_illumina_ncbi_std</code></pre>
</div>
</div>
</section>
</section>
<section id="aln-alignment-files" class="level2">
<h2 class="anchored" data-anchor-id="aln-alignment-files">ALN Alignment Files</h2>
<p>Extension of <a href="https://vtecftwy.github.io/metagentorch/core.html#textfilebasereader"><code>TextFileBaseReader</code></a> class for ALN read/sequence alignment files.</p>
<p>Structure of a ALN sequence file:</p>
<div id="cell-94" class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>p2aln <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads/cov/single_1seq_150bp/single_1seq_150bp.aln'</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> p2aln.is_file()</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>aln <span class="op">=</span> TextFileBaseReader(p2aln, nlines<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, t <span class="kw">in</span> <span class="bu">enumerate</span>(aln):</span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>    txt <span class="op">=</span> t.replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>, <span class="st">''</span>)[:<span class="dv">80</span>]</span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>txt<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb106-8"><a href="#cb106-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">12</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>##ART_Illumina  read_length 150
@CM /usr/bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/ncbi/refs
@SQ 2591237:ncbi:1  1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D    3021
##Header End
&gt;2591237:ncbi:1 2591237:ncbi:1-40200    14370   +
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC
&gt;2591237:ncbi:1 2591237:ncbi:1-40199    15144   -
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGATTCATTTGA
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGA
&gt;2591237:ncbi:1 2591237:ncbi:1-40198    2971    -
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L230" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="alnfilereader" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader">AlnFileReader</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader (path:str|pathlib.Path)</code></pre>
</blockquote>
<p><em>Iterator going through an ALN file</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>path</td>
<td>str | pathlib.Path</td>
<td>path to the aln file</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>key/value with keys:</strong></td>
</tr>
</tbody>
</table>
<div id="cell-96" class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>aln <span class="op">=</span> AlnFileReader(p2aln)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#alnfilereader"><code>AlnFileReader</code></a> iterator returns elements one by one, as dictionaries with each data line related to the read, accessible through the following keys:</p>
<ul>
<li>key <code>'definition line'</code>: <strong>read definition line</strong>, including read metadata</li>
<li>key <code>'ref_seq_aligned'</code>: <strong>aligned reference sequence</strong>, that is the sequence segment in the original reference corresponding to the read</li>
<li>key <code>'read_seq_aligned'</code>: <strong>aligned read</strong>, that is the simmulated read sequence, where each bp corresponds to the reference sequence bp in the same position.</li>
</ul>
<div id="cell-98" class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>one_iteration <span class="op">=</span> <span class="bu">next</span>(aln)</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>one_iteration.keys()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="108">
<pre><code>dict_keys(['definition line', 'ref_seq_aligned', 'read_seq_aligned'])</code></pre>
</div>
</div>
<div id="cell-99" class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>pprint(one_iteration)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'definition line': '&gt;2591237:ncbi:1\t2591237:ncbi:1-40200\t14370\t+',
 'read_seq_aligned': 'TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAGTTTCAAGGAACTTTTAGTGTATGCTGCTGATCCAGCCATGCATGCAGCTT',
 'ref_seq_aligned': 'TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAGTTTCAAGGAACTTTTAGTGTATGCTGCTGATCCAGCCATGCATGCAGCTT'}</code></pre>
</div>
</div>
<div id="cell-100" class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>dfn_line, ref_seq_aligned, read_seq_aligned <span class="op">=</span> one_iteration.values()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-101" class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>dfn_line</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="111">
<pre><code>'&gt;2591237:ncbi:1\t2591237:ncbi:1-40200\t14370\t+'</code></pre>
</div>
</div>
<div id="cell-102" class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>ref_seq_aligned[:<span class="dv">100</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="112">
<pre><code>'TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAG'</code></pre>
</div>
</div>
<div id="cell-103" class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>read_seq_aligned[:<span class="dv">100</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="113">
<pre><code>'TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAACTTACATAGCTCGCGTCTCAG'</code></pre>
</div>
</div>
<div id="cell-104" class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>another_iteration <span class="op">=</span> <span class="bu">next</span>(aln)</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>pprint(another_iteration)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'definition line': '&gt;2591237:ncbi:1\t2591237:ncbi:1-40199\t15144\t-',
 'read_seq_aligned': 'TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGAGTTATAGTAGGGATGACATTACGCTTAGTATACGCGAAAAGTGCATCTTGATCCTCATAACTCATTGAGT',
 'ref_seq_aligned': 'TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGATTCATTTGAGTTATAGTAGGGATGACATTACGCTTAGTATACGCGAAAAGTGCATCTTGATCCTCATAACTCATTGAGT'}</code></pre>
</div>
</div>
<div id="cell-105" class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>aln.reset_iterator()</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, d <span class="kw">in</span> <span class="bu">enumerate</span>(aln):</span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(d[<span class="st">'definition line'</span>])</span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(d[<span class="st">'ref_seq_aligned'</span>][:<span class="dv">80</span>], <span class="st">'...'</span>)</span>
<span id="cb123-5"><a href="#cb123-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(d[<span class="st">'read_seq_aligned'</span>][:<span class="dv">80</span>], <span class="st">'...</span><span class="ch">\n</span><span class="st">'</span>)</span>
<span id="cb123-6"><a href="#cb123-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">3</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;2591237:ncbi:1 2591237:ncbi:1-40200    14370   +
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC ...
TTGTAGATGGTGTTCCTTTTGTTGTTTCAACTGGATACCATTTTCGTGAGTTAGGAGTTGTACATAATCAGGATGTAAAC ...

&gt;2591237:ncbi:1 2591237:ncbi:1-40199    15144   -
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGATTCATTTGA ...
TCATAGTACTACAGATAGAGACACCAGCTACGGTGCGAGCTCTATTCTTTGCACTAATGGCGTACTTAAGAGTCATTTGA ...

&gt;2591237:ncbi:1 2591237:ncbi:1-40198    2971    -
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT ...
TAACATAGTGGTTCGTTTATCAAGGATAATCTATCTCCATAGGTTCTTCATCATCTAACTCTGAATATTTATTCTTAGTT ...

&gt;2591237:ncbi:1 2591237:ncbi:1-40197    15485   -
TAATCACTGATAGCAGCATTGCCATCCTGAGCAAAGAAGAAGTGTTTTAGTTCAACAGAACTTCCTTCCTTAAAGAAACC ...
TAATCACTGATAGCAGCATTGCCATCCTGAGCAAAGAAGAAGTGTTTTAGTTCAACAGAACTTCCTTCCTTAAAGAAACC ...
</code></pre>
</div>
</div>
<p>Once instantiated, the <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#alnfilereader"><code>AlnFileReader</code></a> iterator gives access to the file’s header information through <code>header</code> instance attribute. It is a dictionary with two keys: <code>'command'</code> and <code>'reference sequences'</code>:</p>
<pre><code>    {'command':             'art-illumina command used to create the reads',
     'reference sequences': ['@SQ metadata on reference sequence 1 used for the reads',
                             '@SQ metadata on reference sequence 2 used for the reads', 
                             ...
                            ]
    }</code></pre>
<div id="cell-107" class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(aln.header[<span class="st">'command'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/usr/bin/art_illumina -i /home/vtec/projects/bio/metagentools/data/ncbi/refsequences/cov/cov_refseq_001-seq1.fa -ss HS25 -l 150 -f 200 -o /home/vtec/projects/bio/metagentools/data/ncbi/simreads/cov/single_1seq_150bp/single_1seq_150bp -rs 1723893089</code></pre>
</div>
</div>
<div id="cell-108" class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> seq_info <span class="kw">in</span> aln.header[<span class="st">'reference sequences'</span>]:</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(seq_info)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>@SQ 2591237:ncbi:1  1   MK211378    2591237 ncbi    Coronavirus BtRs-BetaCoV/YN2018D    30213</code></pre>
</div>
</div>
<p>The <strong>read definition line</strong> includes key metadata, which need to be parsed using the appropriate parsing rule.</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/core.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="textfilebasereader.parse_text-1" class="level3">
<h3 class="anchored" data-anchor-id="textfilebasereader.parse_text-1">TextFileBaseReader.parse_text</h3>
<blockquote class="blockquote">
<pre><code> TextFileBaseReader.parse_text (txt:str, pattern:str|None=None)</code></pre>
</blockquote>
<p><em>Parse text using regex pattern with groups. Return a metadata dictionary.</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>txt</td>
<td>str</td>
<td></td>
<td>text to parse</td>
</tr>
<tr class="even">
<td>pattern</td>
<td>str | None</td>
<td>None</td>
<td>If None, uses standard regex pattern to extract metadata, otherwise, uses passed regex</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td><strong>parsed metadata in key/value format</strong></td>
</tr>
</tbody>
</table>
<div id="cell-111" class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>aln.parse_text(dfn_line, pattern)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="120">
<pre><code>{'refseqid': '2591237:ncbi:1',
 'reftaxonomyid': '2591237',
 'refsource': 'ncbi',
 'refseqnb': '1',
 'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'aln_start_pos': '14370',
 'refseq_strand': '+'}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L298" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="alnfilereader.parse_definition_line_with_position" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader.parse_definition_line_with_position">AlnFileReader.parse_definition_line_with_position</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader.parse_definition_line_with_position (dfn_line:str)</code></pre>
</blockquote>
<p><em>Parse definition line and adds relative position</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dfn_line</td>
<td>str</td>
<td>fefinition line string to be parsed</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td><strong>parsed metadata in key/value format + relative position of the read</strong></td>
</tr>
</tbody>
</table>
<p>Upon instance creation, <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#alnfilereader"><code>AlnFileReader</code></a> automatically checks the <code>default_parsing_rules.json</code> file for a workable rule among saved rules. Saved rules include the rule for ART Illumina ALN files.</p>
<div id="cell-114" class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>aln.re_rule_name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="122">
<pre><code>'aln_art_illumina_ncbi_std'</code></pre>
</div>
</div>
<p>It is therefore not required to pass a specific <code>pattern</code> and <code>keys</code> parameter.</p>
<div id="cell-116" class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>aln.parse_text(dfn_line)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="123">
<pre><code>{'refseqid': '2591237:ncbi:1',
 'reftaxonomyid': '2591237',
 'refsource': 'ncbi',
 'refseqnb': '1',
 'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'aln_start_pos': '14370',
 'refseq_strand': '+'}</code></pre>
</div>
</div>
<p>ART Ilumina ALN files definition lines consist of:</p>
<ul>
<li>The <strong>read</strong> ID: <code>readid</code>, e.g.&nbsp;<code>2591237:ncbi:1-20100</code></li>
<li>the <strong>read</strong> number (order in the file): <code>readnb</code>, e.g.&nbsp;<code>20100</code></li>
<li>The <strong>read</strong> start position in the reference sequence: <code>aln_start_pos</code>, e.g.&nbsp;<code>23878</code></li>
<li>The <strong>reference sequence</strong> ID: <code>readid</code>, e.g.&nbsp;<code>2591237:ncbi:1-20100</code></li>
<li>The <strong>reference sequence</strong> number: <code>refseqnb</code>, e.g.&nbsp;<code>1</code></li>
<li>The <strong>reference sequence</strong> source: <code>refsource</code>, e.g.&nbsp;<code>ncbi</code></li>
<li>The <strong>reference sequence</strong> taxonomy: <code>reftaxonomyid</code>, e.g.&nbsp;<code>2591237</code></li>
<li>The <strong>reference sequence</strong> strand: <code>refseq_strand</code> wich is either <code>+</code> or <code>-</code>,</li>
</ul>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L310" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="alnfilereader.parse_file" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader.parse_file">AlnFileReader.parse_file</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader.parse_file (add_ref_seq_aligned:bool=False,
                           add_read_seq_aligned:bool=False)</code></pre>
</blockquote>
<p><em>Read ALN file, return a dict w/ alignment info for each read and optionaly aligned reference sequence &amp; read</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>add_ref_seq_aligned</td>
<td>bool</td>
<td>False</td>
<td>Add the reference sequence aligned to the parsed dictionary when True</td>
</tr>
<tr class="even">
<td>add_read_seq_aligned</td>
<td>bool</td>
<td>False</td>
<td>Add the read sequence aligned to the parsed dictionary when True</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-119" class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>parsed <span class="op">=</span> aln.parse_file()</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (k, v) <span class="kw">in</span> <span class="bu">enumerate</span>(parsed.items()):</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(k)</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>    pprint(v)</span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;</span> <span class="dv">3</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>2591237:ncbi:1-40200
{'aln_start_pos': '14370',
 'readid': '2591237:ncbi:1-40200',
 'readnb': '40200',
 'refseq_strand': '+',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40199
{'aln_start_pos': '15144',
 'readid': '2591237:ncbi:1-40199',
 'readnb': '40199',
 'refseq_strand': '-',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40198
{'aln_start_pos': '2971',
 'readid': '2591237:ncbi:1-40198',
 'readnb': '40198',
 'refseq_strand': '-',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40197
{'aln_start_pos': '15485',
 'readid': '2591237:ncbi:1-40197',
 'readnb': '40197',
 'refseq_strand': '-',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}
2591237:ncbi:1-40196
{'aln_start_pos': '16221',
 'readid': '2591237:ncbi:1-40196',
 'readnb': '40196',
 'refseq_strand': '-',
 'refseqid': '2591237:ncbi:1',
 'refseqnb': '1',
 'refsource': 'ncbi',
 'reftaxonomyid': '2591237'}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L330" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="alnfilereader.parse_header_reference_sequences" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader.parse_header_reference_sequences">AlnFileReader.parse_header_reference_sequences</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader.parse_header_reference_sequences (pattern:str|None=None)</code></pre>
</blockquote>
<p><em>Extract metadata from all header reference sequences</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pattern</td>
<td>str | None</td>
<td>None</td>
<td>regex pattern to apply to parse the reference sequence info</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>dict</strong></td>
<td></td>
<td><strong>parsed metadata in key/value format</strong></td>
</tr>
</tbody>
</table>
<div id="cell-121" class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>pprint(aln.parse_header_reference_sequences())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'2591237:ncbi:1': {'organism': 'Coronavirus BtRs-BetaCoV/YN2018D',
                    'refseq_accession': 'MK211378',
                    'refseq_length': '30213',
                    'refseqid': '2591237:ncbi:1',
                    'refseqnb': '1',
                    'refsource': 'ncbi',
                    'reftaxonomyid': '2591237'}}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L345" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="alnfilereader.set_header_parsing_rules" class="level3">
<h3 class="anchored" data-anchor-id="alnfilereader.set_header_parsing_rules">AlnFileReader.set_header_parsing_rules</h3>
<blockquote class="blockquote">
<pre><code> AlnFileReader.set_header_parsing_rules (pattern:str|None=None,
                                         verbose:bool=False)</code></pre>
</blockquote>
<p>*Set the regex parsing rule for reference sequence in ALN header.</p>
<p>Updates 3 class attributes: <code>re_header_rule_name</code>, <code>re_header_pattern</code>, <code>re_header_keys</code></p>
<p>TODO: refactor this and the method in Core: to use a single function for the common part and a parameter for the text_to_parse*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pattern</td>
<td>str | None</td>
<td>None</td>
<td>regex pattern to apply to parse the text, search in parsing rules json if None</td>
</tr>
<tr class="even">
<td>verbose</td>
<td>bool</td>
<td>False</td>
<td>when True, provides information on each rule</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-123" class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>aln.set_header_parsing_rules(verbose<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--------------------------------------------------------------------------------
Rule &lt;fasta_ncbi_std&gt; generated 0 matches
--------------------------------------------------------------------------------
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*(?P=seqnb)[\s\t](?P&lt;accession&gt;[\w\d]*)([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t][\s\t]*(?P&lt;organism&gt;[\w\s\-\_/]*))?
['seqid', 'taxonomyid', 'source', 'accession', 'seqnb', 'organism']
{'seqid': '', 'taxonomyid': '', 'source': '', 'seqnb': '', 'accession': '', 'organism': ''}
--------------------------------------------------------------------------------
Rule &lt;fastq_art_illumina_ncbi_std&gt; generated 0 matches
--------------------------------------------------------------------------------
^@(?P&lt;readid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*)-(?P&lt;readnb&gt;\d*(\/\d)?))$
['readid', 'reftaxonomyid', 'refsource', 'refseqnb', 'readnb']
{'readid': '', 'reftaxonomyid': '', 'refsource': '', 'refseqnb': '', 'readnb': ''}
--------------------------------------------------------------------------------
Rule &lt;aln_art_illumina_ncbi_std&gt; generated 0 matches
--------------------------------------------------------------------------------
^&gt;(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))(\s| )*(?P&lt;readid&gt;(?P=reftaxonomyid):(?P=refsource):(?P=refseqnb)-(?P&lt;readnb&gt;\d*(\/\d(-\d)?)?))(\s|\t)(?P&lt;aln_start_pos&gt;\d*)(\s|\t)(?P&lt;refseq_strand&gt;(-|\+))$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'readid', 'readnb', 'aln_start_pos', 'refseq_strand']
{'refseqid': '', 'reftaxonomyid': '', 'refsource': '', 'refseqnb': '', 'readid': '', 'readnb': '', 'aln_start_pos': '', 'refseq_strand': ''}
--------------------------------------------------------------------------------
Rule &lt;aln_art_illumina-refseq-ncbi-std&gt; generated 7 matches
--------------------------------------------------------------------------------
^@SQ[\t\s]*(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))[\t\s]*(?P=refseqnb)[\t\s]*(?P&lt;refseq_accession&gt;[\w\d]*)[\t\s]*(?P=reftaxonomyid)[\t\s]*(?P=refsource)[\t\s](?P&lt;organism&gt;.*)[\t\s](?P&lt;refseq_length&gt;\d*)$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'refseq_accession', 'organism', 'refseq_length']
{'refseqid': '2591237:ncbi:1', 'reftaxonomyid': '2591237', 'refsource': 'ncbi', 'refseqnb': '1', 'refseq_accession': 'MK211378', 'organism': 'Coronavirus BtRs-BetaCoV/YN2018D', 'refseq_length': '30213'}
--------------------------------------------------------------------------------
Rule &lt;fasta_ncbi_cov&gt; generated 0 matches
--------------------------------------------------------------------------------
^&gt;(?P&lt;seqid&gt;(?P&lt;taxonomyid&gt;\d+):(?P&lt;source&gt;ncbi):(?P&lt;seqnb&gt;\d*))[\s\t]*\[(?P&lt;accession&gt;[\w\d]*)\]([\s\t]*(?P=taxonomyid)[\s\t]*(?P=source)[\s\t]*(?P=seqnb)[\s\t]*\[(?P=accession)\][\s\t]*(?P=taxonomyid)[\s\t]*(?P&lt;organism&gt;[\w\s\-\_\/]*))?
['seqid', 'taxonomyid', 'source', 'accession', 'seqnb', 'organism']
{'seqid': '', 'taxonomyid': '', 'source': '', 'seqnb': '', 'accession': '', 'organism': ''}
--------------------------------------------------------------------------------
Rule &lt;fasta_rhinolophus_ferrumequinum&gt; generated 0 matches
--------------------------------------------------------------------------------
^&gt;\d[\s\t](?P&lt;seq_type&gt;dna_rm):(?P&lt;id_type&gt;[\w\_]*)[\s\w](?P=id_type):(?P&lt;assy&gt;[\w\d\_]*)\.(?P&lt;seq_level&gt;[\w]*):\d*:\d*:(?P&lt;taxonomy&gt;\d*):(?P&lt;id&gt;\d*)[\s    ]REF$
['seq_type', 'id_type', 'assy', 'seq_level', 'taxonomy', 'id']
{'seq_type': '', 'id_type': '', 'assy': '', 'seq_level': '', 'taxonomy': '', 'id': ''}
--------------------------------------------------------------------------------
Selected rule with most matches: aln_art_illumina-refseq-ncbi-std</code></pre>
</div>
</div>
<div id="cell-124" class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(aln.re_header_rule_name)</span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(aln.re_header_pattern)</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(aln.re_header_keys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>aln_art_illumina-refseq-ncbi-std
^@SQ[\t\s]*(?P&lt;refseqid&gt;(?P&lt;reftaxonomyid&gt;\d*):(?P&lt;refsource&gt;\w*):(?P&lt;refseqnb&gt;\d*))[\t\s]*(?P=refseqnb)[\t\s]*(?P&lt;refseq_accession&gt;[\w\d]*)[\t\s]*(?P=reftaxonomyid)[\t\s]*(?P=refsource)[\t\s](?P&lt;organism&gt;.*)[\t\s](?P&lt;refseq_length&gt;\d*)$
['refseqid', 'reftaxonomyid', 'refsource', 'refseqnb', 'refseq_accession', 'organism', 'refseq_length']</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="build-datasets" class="level1">
<h1>Build datasets</h1>
<p>Reads are provided in various formats (plain text for original data, fastq + aln for simulated reads). The CNN_Virus model requires inputs in the form of three tensors:</p>
<ul>
<li><code>read_seq_batch</code>: a batch of 50-mer read sequences, in “base-hot-encoded” format (BHE)</li>
<li><code>label_batch</code>: a batch of the species’ labels, in one-hot-encoded format (187 classes) (OHE)</li>
<li><code>pos_batch</code>: a batch of the read’s relative positions in the reference sequence, in one-hot-encoded format (10 classes) (OHE)</li>
</ul>
<p>The classes below are custom pytorch <code>Dataset</code> used to load the reads from their file and transform it into BHE or OHE tensors.</p>
<section id="plain-text-based-data" class="level2">
<h2 class="anchored" data-anchor-id="plain-text-based-data">Plain text based data</h2>
<p>The datasets provided by CNN Virus team are in plain text format, one sequence per line, with format as follows:</p>
<pre class="ascii"><code>AAAAAGATTTTGAGAGAGGTCGACCTGTCCTCCTAAAACGTTTACAAAAG  71  0
CATGTAACGCAGCTTAGTCCGATCGTGGCTATAATCCGTCTTTCGATTTG  1   7
AACAACATCTTGTTGATGATAACCGTCAAAGTGTTTTGGGTCTGGAGGGA  158 6
AGTACCTGGAGAGCGTTAAGAAACACAAACGGCTGGATGTAGTGCCGCGC  6   7
CCACGTCGATGAAGCTCCGACGAGAGTCGGCGCTGAGCCCGCGCACCTCC  71  6
AGCTCGTGGATCTCCCCTCCTTCTGCAGTTTCAACATCAGAAGCCCTGAA  87  1</code></pre>
<p>The first column is the k-base (k-mer) read, followed byt the specie label and the relative position.</p>
<p>When using this type of data, the data pipeline consist of:</p>
<ul>
<li><p>Creating a <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#textfiledataset"><code>TextFileDataset</code></a> instance to load the data</p></li>
<li><p>Using the created dataset to in a <code>Dataloader</code> used for training or inference.</p></li>
</ul>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L427" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="textfiledataset" class="level3">
<h3 class="anchored" data-anchor-id="textfiledataset">TextFileDataset</h3>
<blockquote class="blockquote">
<pre><code> TextFileDataset (p2file:str|pathlib.Path)</code></pre>
</blockquote>
<p><em>Load data from text file and yield (BHE sequence tensor, (label OHE tensor, position OHE tensor))</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p2file</td>
<td>str | pathlib.Path</td>
<td>path to the file to read</td>
</tr>
</tbody>
</table>
<p>This <code>IterableDataset</code> reads the file line by line and yields a tupple <code>(seq_bhe, (lbl_ohe, pos_ohe))</code>:</p>
<ul>
<li><code>seq_bhe</code>: tensor of shape [k, 5] with the k-mer read in base-hot-encoded format</li>
<li><code>lbl_ohe</code>: tensor of shape [187] with the specie label in one-hot-encoded format</li>
<li><code>pos_ohe</code>: tensor of shape [10] with the relative position label in one-hot-encoded format</li>
</ul>
<p>The dataset is then used with a pytorch <code>DataLoader</code> to handle batching. When using the GPU, the <code>pin_memory=True</code> should be used to make transfer of data to the GPU faster.</p>
<p>Let’s use a small data file to illustrate the process.</p>
<div id="cell-132" class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a>p2file <span class="op">=</span> Path(<span class="st">'data_dev/CNN_Virus_data/50mer_ds_100_seq'</span>)</span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> p2file.is_file()</span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>ds <span class="op">=</span> TextFileDataset(p2file)</span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a>dl <span class="op">=</span> DataLoader(ds, batch_size<span class="op">=</span><span class="dv">8</span>, num_workers<span class="op">=</span><span class="dv">2</span>, pin_memory<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> seq_batch, (lbl_batch, pos_batch) <span class="kw">in</span> dl:</span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(seq_batch.shape, lbl_batch.shape, pos_batch.shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([8, 50, 5]) torch.Size([8, 187]) torch.Size([8, 10])
torch.Size([4, 50, 5]) torch.Size([4, 187]) torch.Size([4, 10])
torch.Size([4, 50, 5]) torch.Size([4, 187]) torch.Size([4, 10])</code></pre>
</div>
</div>
<p>The read sequence tensor is BHE:</p>
<div id="cell-134" class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a>seq_batch[:<span class="dv">2</span>, :<span class="dv">3</span>, :]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>tensor([[[1, 0, 0, 0, 0],
         [1, 0, 0, 0, 0],
         [0, 1, 0, 0, 0]],

        [[1, 0, 0, 0, 0],
         [0, 0, 0, 1, 0],
         [1, 0, 0, 0, 0]]])</code></pre>
</div>
</div>
<div id="cell-135" class="cell">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>lbl_batch[:<span class="dv">2</span>, :], lbl_batch.argmax(dim<span class="op">=</span><span class="dv">1</span>)[:<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>(tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0.]]),
 tensor([94, 18]))</code></pre>
</div>
</div>
</section>
</section>
<section id="aln-file-based" class="level2">
<h2 class="anchored" data-anchor-id="aln-file-based">ALN file based</h2>
<p>All simulated reads are stored in a <code>.fastq</code> and <code>.aln</code> file. We use the <code>.aln</code> file because it includes all metadata information required to retrieve both the specie and the relative position of each read.</p>
<p><a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#alnfiledataset"><code>AlnFileDataset</code></a> allows to load reads and metadata stored in ART Illumina simulated ALN file.</p>
<div id="cell-138" class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>p2aln <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads/cov/single_1seq_150bp/single_1seq_150bp.aln'</span></span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>aln <span class="op">=</span> AlnFileReader(p2aln)</span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, d <span class="kw">in</span> <span class="bu">enumerate</span>(aln):</span>
<span id="cb157-5"><a href="#cb157-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">pass</span></span>
<span id="cb157-6"><a href="#cb157-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">:,d}</span><span class="ss"> reads in this ALN file"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>40,200 reads in this ALN file</code></pre>
</div>
</div>
<div id="cell-139" class="cell">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a>p2aln <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads/cov/single_1seq_150bp/single_2seq_150bp.aln'</span></span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a>aln.reset_iterator()</span>
<span id="cb159-3"><a href="#cb159-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, seq <span class="kw">in</span> <span class="bu">enumerate</span>(aln):</span>
<span id="cb159-4"><a href="#cb159-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(aln.parse_definition_line_with_position(seq[<span class="st">'definition line'</span>]))</span>
<span id="cb159-5"><a href="#cb159-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i<span class="op">+</span><span class="dv">1</span> <span class="op">&gt;=</span> <span class="dv">3</span>: <span class="cf">break</span></span>
<span id="cb159-6"><a href="#cb159-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb159-7"><a href="#cb159-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, seq <span class="kw">in</span> <span class="bu">enumerate</span>(aln):</span>
<span id="cb159-8"><a href="#cb159-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(aln.parse_definition_line_with_position(seq[<span class="st">'definition line'</span>]))</span>
<span id="cb159-9"><a href="#cb159-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i<span class="op">+</span><span class="dv">1</span> <span class="op">&gt;=</span> <span class="dv">3</span>: <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'refseqid': '2591237:ncbi:1', 'reftaxonomyid': '2591237', 'refsource': 'ncbi', 'refseqnb': '1', 'readid': '2591237:ncbi:1-40200', 'readnb': '40200', 'aln_start_pos': '14370', 'refseq_strand': '+', 'read_pos': 5}
{'refseqid': '2591237:ncbi:1', 'reftaxonomyid': '2591237', 'refsource': 'ncbi', 'refseqnb': '1', 'readid': '2591237:ncbi:1-40199', 'readnb': '40199', 'aln_start_pos': '15144', 'refseq_strand': '-', 'read_pos': 6}
{'refseqid': '2591237:ncbi:1', 'reftaxonomyid': '2591237', 'refsource': 'ncbi', 'refseqnb': '1', 'readid': '2591237:ncbi:1-40198', 'readnb': '40198', 'aln_start_pos': '2971', 'refseq_strand': '-', 'read_pos': 1}
{'refseqid': '2591237:ncbi:1', 'reftaxonomyid': '2591237', 'refsource': 'ncbi', 'refseqnb': '1', 'readid': '2591237:ncbi:1-40197', 'readnb': '40197', 'aln_start_pos': '15485', 'refseq_strand': '-', 'read_pos': 6}
{'refseqid': '2591237:ncbi:1', 'reftaxonomyid': '2591237', 'refsource': 'ncbi', 'refseqnb': '1', 'readid': '2591237:ncbi:1-40196', 'readnb': '40196', 'aln_start_pos': '16221', 'refseq_strand': '-', 'read_pos': 6}
{'refseqid': '2591237:ncbi:1', 'reftaxonomyid': '2591237', 'refsource': 'ncbi', 'refseqnb': '1', 'readid': '2591237:ncbi:1-40195', 'readnb': '40195', 'aln_start_pos': '18953', 'refseq_strand': '-', 'read_pos': 7}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L466" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="alnfiledataset" class="level3">
<h3 class="anchored" data-anchor-id="alnfiledataset">AlnFileDataset</h3>
<blockquote class="blockquote">
<pre><code> AlnFileDataset (p2file:str|pathlib.Path, label:int=118,
                 return_metadata:bool=False, refseqid:str|None=None)</code></pre>
</blockquote>
<p>*Load data and metadata from ALN file, yield BHE sequence, OHE label, OHE position tensors + metadata</p>
<p>The iterator yield tupple (read tensor,(label tensor, position tensor)):</p>
<ul>
<li>kmer read tensor in base hot encoded format (shape [k, 5])</li>
<li>label tensor in one hot encoded format (shape [187])</li>
<li>position tensor in one hote encoded format (shape [10])</li>
</ul>
<p>It also optionally returns a dictionary of the read metadata available in the ALN file.*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p2file</td>
<td>str | pathlib.Path</td>
<td></td>
<td>path to the file to read</td>
</tr>
<tr class="even">
<td>label</td>
<td>int</td>
<td>118</td>
<td>label for this batch (assuming all reads are from the same species)</td>
</tr>
<tr class="odd">
<td>return_metadata</td>
<td>bool</td>
<td>False</td>
<td>yield each read metadata as a dictionary when True</td>
</tr>
<tr class="even">
<td>refseqid</td>
<td>str | None</td>
<td>None</td>
<td>refseqid to filter reference sequence (default is no filtering)</td>
</tr>
</tbody>
</table>
<p>Create a dataset from a ALN file, then pass it to the class <code>Dataloader</code> to create batches of data</p>
<div id="cell-142" class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>ds <span class="op">=</span> AlnFileDataset(p2aln, label<span class="op">=</span><span class="dv">118</span>)</span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>dl <span class="op">=</span> DataLoader(ds, batch_size<span class="op">=</span><span class="dv">16</span>, shuffle<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb162-4"><a href="#cb162-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-5"><a href="#cb162-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (s, (lbl,pos)) <span class="kw">in</span> <span class="bu">enumerate</span>(dl):</span>
<span id="cb162-6"><a href="#cb162-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Sample </span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">:000d}</span><span class="ss">:"</span>)</span>
<span id="cb162-7"><a href="#cb162-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"  Shapes of each data tensor:"</span>)</span>
<span id="cb162-8"><a href="#cb162-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"    seq: </span><span class="sc">{</span>s<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">, label: </span><span class="sc">{</span>lbl<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">, pos: </span><span class="sc">{</span>pos<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb162-9"><a href="#cb162-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i<span class="op">+</span><span class="dv">1</span> <span class="op">&gt;=</span> <span class="dv">3</span>:<span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample 1:
  Shapes of each data tensor:
    seq: torch.Size([16, 150, 5]), label: torch.Size([16, 187]), pos: torch.Size([16, 10])
Sample 2:
  Shapes of each data tensor:
    seq: torch.Size([16, 150, 5]), label: torch.Size([16, 187]), pos: torch.Size([16, 10])
Sample 3:
  Shapes of each data tensor:
    seq: torch.Size([16, 150, 5]), label: torch.Size([16, 187]), pos: torch.Size([16, 10])</code></pre>
</div>
</div>
<p>When an <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#alnfiledataset"><code>AlnFileDataset</code></a> instance with <code>return_metadata</code> set as True is passed to <code>DataLoader</code>, the dataloader will yield (read_tensor,(lbl_tensor, pos_tensor), metadata) where:</p>
<ul>
<li>read_tensor is the batch of kmer read tensor in base hot encoded format (shape [bs, k, 5])</li>
<li>lbl_tensor is the batch of label tensor in one hot encoded format (shape [bs, 187])</li>
<li>pos_tensor is the batch of position tensor in one hote encoded format (shape [bs, 10])</li>
<li>metadata is a dictionary of form <code>key:list</code> like:</li>
</ul>
<div class="sourceCode" id="cb164"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'aln_start_pos'</span>: [<span class="st">'14370'</span>, <span class="st">'15144'</span>, <span class="st">'2971'</span>], </span>
<span id="cb164-3"><a href="#cb164-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'readid'</span>: [<span class="st">'2591237:ncbi:1-40200'</span>, <span class="st">'2591237:ncbi:1-40199'</span>, <span class="st">'2591237:ncbi:1-40198'</span>], </span>
<span id="cb164-4"><a href="#cb164-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">'readnb'</span>: [<span class="st">'40200'</span>, <span class="st">'40199'</span>, <span class="st">'40198'</span>], </span>
<span id="cb164-5"><a href="#cb164-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">'refseq_strand'</span>: [<span class="st">'+'</span>, <span class="st">'-'</span>, <span class="st">'-'</span>], </span>
<span id="cb164-6"><a href="#cb164-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">'refseqid'</span>: [<span class="st">'2591237:ncbi:1'</span>, <span class="st">'2591237:ncbi:1'</span>, <span class="st">'2591237:ncbi:1'</span>], </span>
<span id="cb164-7"><a href="#cb164-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">'refseqnb'</span>: [<span class="st">'1'</span>, <span class="st">'1'</span>, <span class="st">'1'</span>], </span>
<span id="cb164-8"><a href="#cb164-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">'refsource'</span>: [<span class="st">'ncbi'</span>, <span class="st">'ncbi'</span>, <span class="st">'ncbi'</span>], </span>
<span id="cb164-9"><a href="#cb164-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">'reftaxonomyid'</span>: [<span class="st">'2591237'</span>, <span class="st">'2591237'</span>, <span class="st">'2591237'</span>], </span>
<span id="cb164-10"><a href="#cb164-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">'read_pos'</span>: tensor([ <span class="dv">5</span>,  <span class="dv">6</span>,  <span class="dv">1</span>])</span>
<span id="cb164-11"><a href="#cb164-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="cell-144" class="cell">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>ds <span class="op">=</span> AlnFileDataset(p2aln, label<span class="op">=</span><span class="dv">118</span>, return_metadata<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a>dl <span class="op">=</span> DataLoader(ds, batch_size<span class="op">=</span><span class="dv">4</span>, shuffle<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb165-3"><a href="#cb165-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-4"><a href="#cb165-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (s, (lbl,pos), d) <span class="kw">in</span> <span class="bu">enumerate</span>(dl):</span>
<span id="cb165-5"><a href="#cb165-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Sample </span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">:000d}</span><span class="ss">:"</span>)</span>
<span id="cb165-6"><a href="#cb165-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"  Shapes of each data tensor:"</span>)</span>
<span id="cb165-7"><a href="#cb165-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"    seq: </span><span class="sc">{</span>s<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">, label: </span><span class="sc">{</span>lbl<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">, pos: </span><span class="sc">{</span>pos<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb165-8"><a href="#cb165-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"  Metadata dictionary for the batch"</span>)</span>
<span id="cb165-9"><a href="#cb165-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"    </span><span class="sc">{</span>d[<span class="st">'readid'</span>]<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>d[<span class="st">'refseqid'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb165-10"><a href="#cb165-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i<span class="op">+</span><span class="dv">1</span> <span class="op">&gt;=</span> <span class="dv">3</span>:<span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample 1:
  Shapes of each data tensor:
    seq: torch.Size([4, 150, 5]), label: torch.Size([4, 187]), pos: torch.Size([4, 10])
  Metadata dictionary for the batch
    ['2591237:ncbi:1-40200', '2591237:ncbi:2-40199', '2591237:ncbi:1-40198', '2591237:ncbi:1-40197'] ['2591237:ncbi:1', '2591237:ncbi:2', '2591237:ncbi:1', '2591237:ncbi:1']
Sample 2:
  Shapes of each data tensor:
    seq: torch.Size([4, 150, 5]), label: torch.Size([4, 187]), pos: torch.Size([4, 10])
  Metadata dictionary for the batch
    ['2591237:ncbi:2-40196', '2591237:ncbi:1-40195', '2591237:ncbi:2-40194', '2591237:ncbi:2-40193'] ['2591237:ncbi:2', '2591237:ncbi:1', '2591237:ncbi:2', '2591237:ncbi:2']
Sample 3:
  Shapes of each data tensor:
    seq: torch.Size([4, 150, 5]), label: torch.Size([4, 187]), pos: torch.Size([4, 10])
  Metadata dictionary for the batch
    ['2591237:ncbi:1-40192', '2591237:ncbi:1-40191', '2591237:ncbi:1-40190', '2591237:ncbi:1-40189'] ['2591237:ncbi:1', '2591237:ncbi:1', '2591237:ncbi:1', '2591237:ncbi:1']</code></pre>
</div>
</div>
<p><a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#alnfiledataset"><code>AlnFileDataset</code></a> also allows to filter reads to match one specific reference sequence. If this option is selected, only simulated reads generated from that specific reference sequence will be loaded and yielded.</p>
<div id="cell-146" class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a>p2aln_2seq <span class="op">=</span> pfs.data <span class="op">/</span> <span class="st">'ncbi/simreads/cov/single_1seq_150bp/single_2seq_150bp.aln'</span></span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> p2aln_2seq.is_file()</span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-4"><a href="#cb167-4" aria-hidden="true" tabindex="-1"></a>ds <span class="op">=</span> AlnFileDataset(p2aln_2seq, label<span class="op">=</span><span class="dv">118</span>, return_metadata<span class="op">=</span><span class="va">True</span>, refseqid<span class="op">=</span><span class="st">'2591237:ncbi:2'</span>)</span>
<span id="cb167-5"><a href="#cb167-5" aria-hidden="true" tabindex="-1"></a>dl <span class="op">=</span> DataLoader(ds, batch_size<span class="op">=</span><span class="dv">3</span>, shuffle<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb167-6"><a href="#cb167-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-7"><a href="#cb167-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (s, (lbl,pos), d) <span class="kw">in</span> <span class="bu">enumerate</span>(dl):</span>
<span id="cb167-8"><a href="#cb167-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Sample </span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">:000d}</span><span class="ss">:"</span>)</span>
<span id="cb167-9"><a href="#cb167-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"  Shapes of each data tensor:"</span>)</span>
<span id="cb167-10"><a href="#cb167-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"    seq: </span><span class="sc">{</span>s<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">, label: </span><span class="sc">{</span>lbl<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">, pos: </span><span class="sc">{</span>pos<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb167-11"><a href="#cb167-11" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"  Metadata dictionary for the batch"</span>)</span>
<span id="cb167-12"><a href="#cb167-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"    </span><span class="sc">{</span>d[<span class="st">'readid'</span>]<span class="sc">}</span><span class="ss"> from </span><span class="sc">{</span>d[<span class="st">'refseqid'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb167-13"><a href="#cb167-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i<span class="op">+</span><span class="dv">1</span> <span class="op">&gt;=</span> <span class="dv">3</span>: </span>
<span id="cb167-14"><a href="#cb167-14" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"Breaking out of the loop"</span>)</span>
<span id="cb167-15"><a href="#cb167-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtering reads to only keep those aligned to refseqid 2591237:ncbi:2
Sample 1:
  Shapes of each data tensor:
    seq: torch.Size([3, 150, 5]), label: torch.Size([3, 187]), pos: torch.Size([3, 10])
  Metadata dictionary for the batch
    ['2591237:ncbi:2-40199', '2591237:ncbi:2-40196', '2591237:ncbi:2-40194'] from ['2591237:ncbi:2', '2591237:ncbi:2', '2591237:ncbi:2']
Sample 2:
  Shapes of each data tensor:
    seq: torch.Size([3, 150, 5]), label: torch.Size([3, 187]), pos: torch.Size([3, 10])
  Metadata dictionary for the batch
    ['2591237:ncbi:2-40193', '2591237:ncbi:2-40185', '2591237:ncbi:2-40183'] from ['2591237:ncbi:2', '2591237:ncbi:2', '2591237:ncbi:2']
Sample 3:
  Shapes of each data tensor:
    seq: torch.Size([1, 150, 5]), label: torch.Size([1, 187]), pos: torch.Size([1, 10])
  Metadata dictionary for the batch
    ['2591237:ncbi:2-30794'] from ['2591237:ncbi:2']
Breaking out of the loop</code></pre>
</div>
</div>
</section>
</section>
<section id="handle-long-reads" class="level2">
<h2 class="anchored" data-anchor-id="handle-long-reads">Handle long reads</h2>
<section id="preprocessing" class="level3">
<h3 class="anchored" data-anchor-id="preprocessing">Preprocessing</h3>
<p>CNN Virus handles 50-mer reads only. Longer reads need to be split into 50-mer reads by sliding a window of size 50 along the read sequence. The set of multiple 50-mer reads is then used as input to the model. After prediction on the set of 50-mer reads, the final prediction is obtained by filtering 50-mer prediction that yield a hight enough probability and then voting for the most predicted species.</p>
<p>A k-mer (150-mer) read will be split in k-49 (101) 50-mer reads which will yield k-49 (101) probability tensors. The final prediction will be the species with the highest number of votes, as long as their respective probabilities &gt; 90%.</p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L520" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="split_kmer_batch_into_50mers" class="level3">
<h3 class="anchored" data-anchor-id="split_kmer_batch_into_50mers">split_kmer_batch_into_50mers</h3>
<blockquote class="blockquote">
<pre><code> split_kmer_batch_into_50mers (kmer_b:torch.Tensor,
                               labels:tuple[torch.Tensor,torch.Tensor]|Non
                               e=None)</code></pre>
</blockquote>
<p>*Convert a batch of k-mer reads into 50-mer reads, by shifting the k-mer one base at a time.</p>
<p>Shapes:</p>
<ul>
<li>kmer_b: [b,k,5] -&gt; [b * (k - 49), 50, 5]</li>
<li>label/position_b: [b, nb_class] -&gt; [b * (k - 49), nb_class]</li>
</ul>
<p>Technical Note: we use advanced indexing of the tensor to create the 50-mer and roll them, with no loop.*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>kmer_b</td>
<td>Tensor</td>
<td></td>
<td>tensor representing a batch of k-mer reads, BHE format, shape [b, k, 5]</td>
</tr>
<tr class="even">
<td>labels</td>
<td>tuple[torch.Tensor, torch.Tensor] | None</td>
<td>None</td>
<td>optional tuple with tensor for label and position batches</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>tuple</strong></td>
<td></td>
<td><strong>batch tensor for 50-mer reads, (optional) labels and positions batch tensor</strong></td>
</tr>
</tbody>
</table>
<p>We can test the function, with a test tensor designed to make the validation easier.</p>
<p>We create a tensor with a batch of b k-mer reads in the following format:</p>
<pre class="ascii"><code>read 1: 10001, 10002, 10003, .... 10149
read 2: 20001, 20002, 20003, .... 20149
read 3: 30001, 30002, 30003, .... 30149
    ...</code></pre>
<p>We then add an additional dimension to simulate the BHE encoding, by repeating the same value 5 times.</p>
<div id="cell-152" class="cell">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a>b, k <span class="op">=</span> <span class="dv">4</span>, <span class="dv">150</span></span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-3"><a href="#cb171-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _create_kmer_batch(b:<span class="bu">int</span>, k:<span class="bu">int</span>)<span class="op">-&gt;</span> torch.Tensor:</span>
<span id="cb171-4"><a href="#cb171-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Create a batch of kmer reads, with a fake encoding in 5 dimensions (like BHE)"""</span></span>
<span id="cb171-5"><a href="#cb171-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Build a tensor of shape (b,k) with the pattern described above</span></span>
<span id="cb171-6"><a href="#cb171-6" aria-hidden="true" tabindex="-1"></a>    kmer_b <span class="op">=</span> torch.stack([(i<span class="op">+</span><span class="dv">1</span>) <span class="op">*</span> <span class="dv">10000</span> <span class="op">+</span> torch.arange(k) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(b)], dim<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb171-7"><a href="#cb171-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># simulate the BHE buy duplicating the 5 dimensions</span></span>
<span id="cb171-8"><a href="#cb171-8" aria-hidden="true" tabindex="-1"></a>    kmer_b <span class="op">=</span> torch.stack([kmer_b]<span class="op">*</span><span class="dv">5</span>, dim<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb171-9"><a href="#cb171-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> kmer_b</span>
<span id="cb171-10"><a href="#cb171-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-11"><a href="#cb171-11" aria-hidden="true" tabindex="-1"></a>kmer_b <span class="op">=</span> _create_kmer_batch(b,k)</span>
<span id="cb171-12"><a href="#cb171-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(kmer_b.shape)</span>
<span id="cb171-13"><a href="#cb171-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Slice the tensor to only show the first and last three bases, and only one of the 5 BHE dimensions</span></span>
<span id="cb171-14"><a href="#cb171-14" aria-hidden="true" tabindex="-1"></a>idx_bases <span class="op">=</span> np.array([<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">147</span>,<span class="dv">148</span>,<span class="dv">149</span>])</span>
<span id="cb171-15"><a href="#cb171-15" aria-hidden="true" tabindex="-1"></a>kmer_b[:, idx_bases, <span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>torch.Size([4, 150, 5])</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>tensor([[10000, 10001, 10002, 10147, 10148, 10149],
        [20000, 20001, 20002, 20147, 20148, 20149],
        [30000, 30001, 30002, 30147, 30148, 30149],
        [40000, 40001, 40002, 40147, 40148, 40149]])</code></pre>
</div>
</div>
<p>We see that the function <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#split_kmer_batch_into_50mers"><code>split_kmer_batch_into_50mers</code></a> returns a tensor with shape [b, k-49, 50, 5] as expected.</p>
<div id="cell-154" class="cell">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a>split_50mer_read_b, (split_50mer_label_b, split_50mer_position_b) <span class="op">=</span> split_kmer_batch_into_50mers(kmer_b)</span>
<span id="cb174-2"><a href="#cb174-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(kmer_b.shape)</span>
<span id="cb174-3"><a href="#cb174-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(split_50mer_read_b.shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>torch.Size([4, 150, 5])
torch.Size([404, 50, 5])</code></pre>
</div>
</div>
<p>Lets slice only the first and last two 50-mer reads for each batch and a selection of bases to confirm that the batch is properly split.</p>
<div id="cell-156" class="cell">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>dim0_idxs <span class="op">=</span> [<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">99</span>,<span class="dv">100</span>,<span class="dv">101</span>,<span class="dv">102</span>,<span class="dv">200</span>,<span class="dv">201</span>,<span class="dv">202</span>,<span class="dv">203</span>,<span class="dv">301</span>,<span class="dv">302</span>]</span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a>dim1_idxs <span class="op">=</span> [<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">25</span>,<span class="dv">26</span>,<span class="dv">48</span>,<span class="dv">49</span>]</span>
<span id="cb176-3"><a href="#cb176-3" aria-hidden="true" tabindex="-1"></a>dim2_idxs <span class="op">=</span> [<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>]</span>
<span id="cb176-4"><a href="#cb176-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(split_50mer_read_b[np.ix_(dim0_idxs, dim1_idxs, dim2_idxs)][:,:,<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tensor([[10000, 10001, 10025, 10026, 10048, 10049],
        [10001, 10002, 10026, 10027, 10049, 10050],
        [10099, 10100, 10124, 10125, 10147, 10148],
        [10100, 10101, 10125, 10126, 10148, 10149],
        [20000, 20001, 20025, 20026, 20048, 20049],
        [20001, 20002, 20026, 20027, 20049, 20050],
        [20099, 20100, 20124, 20125, 20147, 20148],
        [20100, 20101, 20125, 20126, 20148, 20149],
        [30000, 30001, 30025, 30026, 30048, 30049],
        [30001, 30002, 30026, 30027, 30049, 30050],
        [30099, 30100, 30124, 30125, 30147, 30148],
        [30100, 30101, 30125, 30126, 30148, 30149]])</code></pre>
</div>
</div>
</section>
<section id="postprocessing" class="level3">
<h3 class="anchored" data-anchor-id="postprocessing">Postprocessing</h3>
<p>The model CNN Virus requires post inderence processing of the predictions: - before inference, each k-mer was split into <span class="math inline">\(k-49\)</span> 50-mer, where were presented to the model for inference. Each k-mer led to <span class="math inline">\(k-49\)</span> predictions (probability tensors) - now, we need to combine these $n = k-49 $ probability tensor into a single prediction for the original k-mer. This is done by first filtering all 50-reads that gave a max probability lower than a specific threshold (0.9 by default) and then combining the prediction by a simple vote</p>
<p>Now we can build the function <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#combine_predictions"><code>combine_predictions</code></a> step by step.</p>
<p>First, lets create a test probabilities tensor representing a batch of <code>bs</code> k-mers, each split into <code>k-49</code> 50-mers, with a probability tensor of shape <code>[bs, k-49, nb_class]</code>.</p>
<div id="cell-160" class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_test_probs(bs<span class="op">=</span><span class="dv">4</span>, k<span class="op">=</span><span class="dv">150</span>, nb_class<span class="op">=</span><span class="dv">187</span>):</span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Creating a batch of probabilities for </span><span class="sc">{</span>bs<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>k<span class="sc">}</span><span class="ss">-mer reads with </span><span class="sc">{</span>nb_class<span class="sc">}</span><span class="ss"> classes:"</span>)</span>
<span id="cb178-3"><a href="#cb178-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-4"><a href="#cb178-4" aria-hidden="true" tabindex="-1"></a>    probs <span class="op">=</span> torch.from_numpy(np.random.rand(bs <span class="op">*</span> n ,nb_class)).reshape(<span class="op">-</span><span class="dv">1</span>,n,nb_class)</span>
<span id="cb178-5"><a href="#cb178-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"probs </span><span class="sc">{</span>probs<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb178-6"><a href="#cb178-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> probs</span>
<span id="cb178-7"><a href="#cb178-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-8"><a href="#cb178-8" aria-hidden="true" tabindex="-1"></a>bs, k, nb_class <span class="op">=</span> <span class="dv">4</span>, <span class="dv">55</span>, <span class="dv">10</span></span>
<span id="cb178-9"><a href="#cb178-9" aria-hidden="true" tabindex="-1"></a>n <span class="op">=</span> k <span class="op">-</span> <span class="dv">49</span></span>
<span id="cb178-10"><a href="#cb178-10" aria-hidden="true" tabindex="-1"></a>probs <span class="op">=</span> create_test_probs(bs, k, nb_class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Creating a batch of probabilities for 4 55-mer reads with 10 classes:
probs torch.Size([4, 6, 10])</code></pre>
</div>
</div>
<p><strong>Step 1</strong>: Extract the predictions from probabilities for each 50-mer read, then filter the predictions with a probability lower than a threshold.</p>
<div id="cell-162" class="cell">
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a>threshold <span class="op">=</span> <span class="fl">0.8</span></span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a>preds <span class="op">=</span> probs.argmax(dim<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb180-3"><a href="#cb180-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"preds </span><span class="sc">{</span>preds<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">:</span><span class="ch">\n</span><span class="ss">"</span>,preds)</span>
<span id="cb180-4"><a href="#cb180-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-5"><a href="#cb180-5" aria-hidden="true" tabindex="-1"></a>INVALID <span class="op">=</span> <span class="dv">9999</span></span>
<span id="cb180-6"><a href="#cb180-6" aria-hidden="true" tabindex="-1"></a>invalid_labels_filter <span class="op">=</span> probs.<span class="bu">max</span>(dim<span class="op">=</span><span class="dv">2</span>).values <span class="op">&lt;=</span> threshold</span>
<span id="cb180-7"><a href="#cb180-7" aria-hidden="true" tabindex="-1"></a>preds[invalid_labels_filter] <span class="op">=</span> INVALID</span>
<span id="cb180-8"><a href="#cb180-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"preds </span><span class="sc">{</span>preds<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">:</span><span class="ch">\n</span><span class="ss">"</span>,preds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>preds torch.Size([4, 6]):
 tensor([[6, 6, 3, 7, 6, 8],
        [6, 5, 1, 2, 6, 9],
        [4, 3, 8, 4, 4, 3],
        [4, 5, 8, 0, 6, 6]])
preds torch.Size([4, 6]):
 tensor([[6, 6, 3, 7, 6, 8],
        [6, 5, 1, 2, 6, 9],
        [4, 3, 8, 4, 4, 3],
        [4, 5, 8, 0, 6, 6]])</code></pre>
</div>
</div>
<p><strong>Step 2</strong>: Extract the prediction unique values and their counts.</p>
<blockquote class="blockquote">
<p>Code explanation note:</p>
<p>We do not want to use a python loop. This is why we use <code>torch.unique(preds)</code> with <code>return_inverse</code>. The function returns a flat tensor with the unique values accross the entire <code>preds</code> and a tensor of the same shape as <code>preds</code> with the indices of the unique values in <code>preds</code>.</p>
</blockquote>
<div class="sourceCode" id="cb182"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> torch.tensor([[<span class="dv">10</span>, <span class="dv">30</span>, <span class="dv">20</span>, <span class="dv">30</span>],[<span class="dv">20</span>,<span class="dv">20</span>,<span class="dv">20</span>,<span class="dv">50</span>]], dtype<span class="op">=</span>torch.<span class="bu">long</span>)</span>
<span id="cb182-2"><a href="#cb182-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-3"><a href="#cb182-3" aria-hidden="true" tabindex="-1"></a>    uniques, inverse <span class="op">=</span> torch.unique(x, <span class="bu">sorted</span><span class="op">=</span><span class="va">True</span>, return_inverse<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb182-4"><a href="#cb182-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-5"><a href="#cb182-5" aria-hidden="true" tabindex="-1"></a>    <span class="op">&gt;</span> unique:  tensor([<span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">30</span>, <span class="dv">50</span>])</span>
<span id="cb182-6"><a href="#cb182-6" aria-hidden="true" tabindex="-1"></a>      inverse: tensor([[<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>],</span>
<span id="cb182-7"><a href="#cb182-7" aria-hidden="true" tabindex="-1"></a>                       [<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">3</span>]]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="cell-164" class="cell">
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get unique values and their counts for the entire tensor</span></span>
<span id="cb183-2"><a href="#cb183-2" aria-hidden="true" tabindex="-1"></a>unique_values, inverse_indices <span class="op">=</span> torch.unique(preds, return_inverse<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb183-3"><a href="#cb183-3" aria-hidden="true" tabindex="-1"></a>inverse_indices <span class="op">=</span> inverse_indices.view(preds.shape)</span>
<span id="cb183-4"><a href="#cb183-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"unique_values </span><span class="sc">{</span>unique_values<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">:</span><span class="ch">\n</span><span class="ss">"</span>,unique_values)</span>
<span id="cb183-5"><a href="#cb183-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"inverse_indices </span><span class="sc">{</span>inverse_indices<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">:</span><span class="ch">\n</span><span class="ss">"</span>,inverse_indices)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>unique_values torch.Size([10]):
 tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
inverse_indices torch.Size([4, 6]):
 tensor([[6, 6, 3, 7, 6, 8],
        [6, 5, 1, 2, 6, 9],
        [4, 3, 8, 4, 4, 3],
        [4, 5, 8, 0, 6, 6]])</code></pre>
</div>
</div>
<p><strong>Step 3</strong>: Compute the counts of unique values in each 50-mer read and store it in a tensor of shape (batch size, nb of unique values in the the shole tensor <code>preds</code>).</p>
<div id="cell-166" class="cell">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a tensor to hold the counts (shape (bs, nb_unique_values_across_the_batch))</span></span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a>counts <span class="op">=</span> torch.zeros((preds.shape[<span class="dv">0</span>], unique_values.shape[<span class="dv">0</span>]), dtype<span class="op">=</span>torch.int64)</span>
<span id="cb185-3"><a href="#cb185-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Count occurrences of each unique value per row</span></span>
<span id="cb185-4"><a href="#cb185-4" aria-hidden="true" tabindex="-1"></a>counts.scatter_add_(dim<span class="op">=</span><span class="dv">1</span>, index<span class="op">=</span>inverse_indices, src<span class="op">=</span>torch.ones_like(inverse_indices, dtype<span class="op">=</span>torch.int64))</span>
<span id="cb185-5"><a href="#cb185-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"unique_values </span><span class="sc">{</span>unique_values<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">:</span><span class="ch">\n</span><span class="ss">"</span>, unique_values)</span>
<span id="cb185-6"><a href="#cb185-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"counts </span><span class="sc">{</span>counts<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">:</span><span class="ch">\n</span><span class="ss">"</span>, counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>unique_values torch.Size([10]):
 tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
counts torch.Size([4, 10]):
 tensor([[0, 0, 0, 1, 0, 0, 3, 1, 1, 0],
        [0, 1, 1, 0, 0, 1, 2, 0, 0, 1],
        [0, 0, 0, 2, 3, 0, 0, 0, 1, 0],
        [1, 0, 0, 0, 1, 1, 2, 0, 1, 0]])</code></pre>
</div>
</div>
<p><strong>Step 4</strong>: Get the index of the most frequent value for each 50-mer read, excluding the INVALID placeholder, and extract the corresponding values into a tensor of shape (<code>bs</code>).</p>
<div id="cell-168" class="cell">
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get value most voted per 50-read (vertical tensor), excluding the placeholder INVALID</span></span>
<span id="cb187-2"><a href="#cb187-2" aria-hidden="true" tabindex="-1"></a>most_voted_idxs <span class="op">=</span> counts[:, :<span class="op">-</span><span class="dv">1</span>].argmax(dim<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb187-3"><a href="#cb187-3" aria-hidden="true" tabindex="-1"></a>most_voted_value <span class="op">=</span> unique_values[most_voted_idxs][:, <span class="va">None</span>]</span>
<span id="cb187-4"><a href="#cb187-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"most_voted_value </span><span class="sc">{</span>most_voted_value<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">:</span><span class="ch">\n</span><span class="ss">"</span>, most_voted_value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>most_voted_value torch.Size([4, 1]):
 tensor([[6],
        [6],
        [4],
        [6]])</code></pre>
</div>
</div>
<p>We put it all in the function <a href="https://vtecftwy.github.io/metagentorch/cnn_virus_data.html#combine_predictions"><code>combine_predictions</code></a></p>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L570" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="combine_predictions" class="level3">
<h3 class="anchored" data-anchor-id="combine_predictions">combine_predictions</h3>
<blockquote class="blockquote">
<pre><code> combine_predictions (label_probs:torch.Tensor, pos_probs:torch.Tensor,
                      threshold:float=0.9)</code></pre>
</blockquote>
<p>*Combine a batch of 50-mer probabilities into one batch of final prediction for label and position</p>
<p>Note: the input must be of shape (batch_size, n, c) where n is k-49 and c is the nb of labels or positions*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>label_probs</td>
<td>Tensor</td>
<td></td>
<td>Probabilities for the labels classes for each 50-mer (shape: [bs, k-49,187])</td>
</tr>
<tr class="even">
<td>pos_probs</td>
<td>Tensor</td>
<td></td>
<td>Probabilities for the position classes for each 50-mer (shape: [bs, k-49,10])</td>
</tr>
<tr class="odd">
<td>threshold</td>
<td>float</td>
<td>0.9</td>
<td>Threshold to consider a prediction as valid</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>tuple</strong></td>
<td></td>
<td><strong>Predicted labels and positions for each read (shape: [bs, 187], [bs, 10])</strong></td>
</tr>
</tbody>
</table>
<p>Testing the function with the same test tensor</p>
<div id="cell-172" class="cell">
<div class="sourceCode cell-code" id="cb190"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a>combine_predictions(probs, probs[:, :,:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>(tensor([3, 1, 4, 6]), tensor([3, 1, 4, 6]))</code></pre>
</div>
</div>
<p>Testing the function with test tensor correponding to real probabilities</p>
<div id="cell-174" class="cell">
<div class="sourceCode cell-code" id="cb192"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb192-1"><a href="#cb192-1" aria-hidden="true" tabindex="-1"></a>bs, k <span class="op">=</span> <span class="dv">1</span>, <span class="dv">150</span></span>
<span id="cb192-2"><a href="#cb192-2" aria-hidden="true" tabindex="-1"></a>label_probs <span class="op">=</span> create_test_probs(bs, k, <span class="dv">187</span>)</span>
<span id="cb192-3"><a href="#cb192-3" aria-hidden="true" tabindex="-1"></a>pos_probs <span class="op">=</span> create_test_probs(bs, k, <span class="dv">10</span>)</span>
<span id="cb192-4"><a href="#cb192-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-5"><a href="#cb192-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'</span><span class="ch">\n</span><span class="st">combine for a batch of data'</span>)</span>
<span id="cb192-6"><a href="#cb192-6" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> combine_predictions(label_probs,pos_probs)</span>
<span id="cb192-7"><a href="#cb192-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(p)</span>
<span id="cb192-8"><a href="#cb192-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'</span><span class="ch">\n</span><span class="st">combine for a single sample (not a batch)'</span>)</span>
<span id="cb192-9"><a href="#cb192-9" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> combine_predictions(label_probs[<span class="dv">0</span>,:,:],pos_probs[<span class="dv">0</span>,:,:])</span>
<span id="cb192-10"><a href="#cb192-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Creating a batch of probabilities for 1 150-mer reads with 187 classes:
probs torch.Size([1, 6, 187])
Creating a batch of probabilities for 1 150-mer reads with 10 classes:
probs torch.Size([1, 6, 10])

combine for a batch of data
(tensor(82), tensor(6))

combine for a single sample (not a batch)
Converting probability tensors to 3 dimensions
(tensor(82), tensor(6))</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="deprecated-items" class="level1">
<h1>Deprecated Items</h1>
<p>When any of the following classes and functions is called, it will raise an exception with an error message indicating how to handle the required code refactoring.</p>
<p>Example:</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb194-1"><a href="#cb194-1" aria-hidden="true" tabindex="-1"></a><span class="op">---------------------------------------------------------------------------</span></span>
<span id="cb194-2"><a href="#cb194-2" aria-hidden="true" tabindex="-1"></a><span class="pp">DeprecationWarning</span>                        Traceback (most recent call last)</span>
<span id="cb194-3"><a href="#cb194-3" aria-hidden="true" tabindex="-1"></a>Input In [<span class="dv">484</span>], <span class="kw">in</span> <span class="op">&lt;</span>cell line: <span class="dv">1</span><span class="op">&gt;</span>()</span>
<span id="cb194-4"><a href="#cb194-4" aria-hidden="true" tabindex="-1"></a><span class="op">----&gt;</span> <span class="dv">1</span> combine_prediction_batch(label_probs, pos_probs)</span>
<span id="cb194-5"><a href="#cb194-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-6"><a href="#cb194-6" aria-hidden="true" tabindex="-1"></a>Input In [<span class="dv">481</span>], <span class="kw">in</span> combine_prediction_batch(<span class="op">*</span>args, <span class="op">**</span>kwargs)</span>
<span id="cb194-7"><a href="#cb194-7" aria-hidden="true" tabindex="-1"></a>      <span class="dv">3</span> <span class="st">"""Deprecated"""</span></span>
<span id="cb194-8"><a href="#cb194-8" aria-hidden="true" tabindex="-1"></a>      <span class="dv">4</span> msg <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb194-9"><a href="#cb194-9" aria-hidden="true" tabindex="-1"></a><span class="st">      5 `combine_prediction_batch` is deprecated. </span></span>
<span id="cb194-10"><a href="#cb194-10" aria-hidden="true" tabindex="-1"></a><span class="st">      6 Use `combine_predictions` instead, with same capabilities and more."""</span></span>
<span id="cb194-11"><a href="#cb194-11" aria-hidden="true" tabindex="-1"></a><span class="op">----&gt;</span> <span class="dv">7</span> <span class="cf">raise</span> <span class="pp">DeprecationWarning</span>(msg)</span>
<span id="cb194-12"><a href="#cb194-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-13"><a href="#cb194-13" aria-hidden="true" tabindex="-1"></a><span class="pp">DeprecationWarning</span>: </span>
<span id="cb194-14"><a href="#cb194-14" aria-hidden="true" tabindex="-1"></a>    `combine_prediction_batch` <span class="kw">is</span> deprecated. </span>
<span id="cb194-15"><a href="#cb194-15" aria-hidden="true" tabindex="-1"></a>    Use `combine_predictions` instead, <span class="cf">with</span> same capabilities <span class="kw">and</span> more.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
<p><a href="https://github.com/vtecftwy/metagentorch/blob/main/metagentorch/cnn_virus/data.py#L625" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="combine_prediction_batch" class="level3">
<h3 class="anchored" data-anchor-id="combine_prediction_batch">combine_prediction_batch</h3>
<blockquote class="blockquote">
<pre><code> combine_prediction_batch (*args, **kwargs)</code></pre>
</blockquote>
<p><em>Deprecated</em></p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/vtecftwy\.github\.io\/metagentorch");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/vtecftwy/metagentorch/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>