{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#|default_exp art"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running locally\n"
     ]
    }
   ],
   "source": [
    "#|hide\n",
    "import nbdev\n",
    "\n",
    "# from __future__ import annotations\n",
    "from fastcore.test import *\n",
    "from metagentorch.cnn_virus.utils import setup_nb\n",
    "from nbdev.showdoc import *\n",
    "\n",
    "ON_COLAB, _, _ = setup_nb(_dev=True)\n",
    "\n",
    "assert not ON_COLAB, f\"Art Illumina cannot be run on Colab\"\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#|export\n",
    "# Imports all dependencies\n",
    "import os\n",
    "import subprocess\n",
    "import shlex\n",
    "\n",
    "from fastcore.basics import patch_to\n",
    "from fastcore.utils import run, join_path_file\n",
    "from pathlib import Path\n",
    "from typing import Tuple, List, Optional"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# art\n",
    "\n",
    "> Use **ART**, next gen read simulation tool, from within a python notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`ART` is an open source package simmulation next generation read of genomes, available on the website of the National Institute of Environmental Health Sciences [here](https://www.niehs.nih.gov/research/resources/software/biostatistics/art/index.cfm). It is a command line interface package. This module makes the package accessible from a jupyter notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Typical usage**\n",
    "\n",
    "- read simulation with paired reads:\n",
    "    - `art_illumina -ss HS25 -sam -i file.fa -p -l 150 -f 20 -m 200 -s 10 -o paired_seq_1`\n",
    "- read simulation  with single reads:\n",
    "    - `art_illumina -ss HS25 -sam -i file.fa -l 150 -f 10 -o single_seq_1`\n",
    "\n",
    "Where the parameters are:\n",
    "\n",
    "```ASCII\n",
    "  -f   --fcov   the fold of read coverage to be simulated or number of reads/read pairs generated for each amplicon\n",
    "  -i   --in     the filename of input DNA/RNA reference\n",
    "  -l   --len    the length of reads to be simulated\n",
    "  -m   --mflen  the mean size of DNA/RNA fragments for paired-end simulations\n",
    "  -o   --out    the prefix of output filename\n",
    "  -p   --paired indicate a paired-end read simulation or to generate reads from both ends of amplicons\n",
    "                NOTE: art will automatically switch to a mate-pair simulation if the given mean fragment size >= 2000\n",
    "  -s   --sdev   the standard deviation of DNA/RNA fragment size for paired-end simulations.\n",
    "  -sam --samout indicate to generate SAM alignment file\n",
    "  -ss  --seqSys The name of Illumina sequencing system of the built-in profile used for simulation\n",
    "                NOTE: sequencing system ID names are:\n",
    "                GA1 - GenomeAnalyzer I (36bp,44bp), GA2 - GenomeAnalyzer II (50bp, 75bp)\n",
    "                HS10 - HiSeq 1000 (100bp),          HS20 - HiSeq 2000 (100bp),      HS25 - HiSeq 2500 (125bp, 150bp)\n",
    "                HSXn - HiSeqX PCR free (150bp),     HSXt - HiSeqX TruSeq (150bp),   MinS - MiniSeq TruSeq (50bp)\n",
    "                MSv1 - MiSeq v1 (250bp),            MSv3 - MiSeq v3 (250bp),        NS50 - NextSeq500 v2 (75bp)\n",
    "```\n",
    "Notes:\n",
    "\n",
    "- For single-end simulation, ART requires input sequence file, output file prefix, read length, and read count/fold coverage.\n",
    "- For paired-end simulation (except for amplicon sequencing), ART also requires the parameter values of the mean and standard deviation of DNA/RNA fragment lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "\n",
    "# Private Utility functions to export ==============================================\n",
    "\n",
    "def _run(args:List[str], shell:bool=False, print_output=True):\n",
    "    \"\"\"Wrapper subprocess.run and prints the output\"\"\"\n",
    "    p = subprocess.run(args=args, stdout=subprocess.PIPE, shell=shell)\n",
    "    if print_output:\n",
    "        print('return code: ',p.returncode, '\\n')\n",
    "        print(str(p.stdout, 'utf-8'))\n",
    "\n",
    "def _validate_path(p:str|Path) -> Path:\n",
    "    \"\"\"checks that path is a string or a Path, and returns a Path\"\"\"\n",
    "    if isinstance(p, str): \n",
    "        p = Path(p)\n",
    "    elif not isinstance(p, Path): \n",
    "        raise TypeError(f\"a path must be a string or a Path, not a {type(p)}\")\n",
    "    return p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "class ArtIllumina:\n",
    "    \"\"\"Class to handle all aspects of simulating sequencing with art_illumina\"\"\"\n",
    "\n",
    "    QUALITY_PROFILES = {\n",
    "        'GA1': 'GenomeAnalyzer I (36bp,44bp)', 'GA2': 'GenomeAnalyzer II (50bp, 75bp)',\n",
    "        'HS10': 'HiSeq 1000 (100bp)', 'HS20': 'HiSeq 2000 (100bp)',\n",
    "        'HS25': 'HiSeq 2500 (125bp, 150bp)', 'HSXn': 'HiSeqX PCR free (150bp)', 'HSXt': 'HiSeqX TruSeq (150bp)',\n",
    "        'MinS': 'MiniSeq TruSeq (50bp)', 'MSv1': 'MiSeq v1 (250bp)', 'MSv3': 'MiSeq v3 (250bp)',\n",
    "        'NS50': 'NextSeq500 v2 (75bp)'\n",
    "        }\n",
    "\n",
    "    def __init__(\n",
    "        self, \n",
    "        path2app: str|Path,            # full path to art_illumina application on the system\n",
    "        input_dir: str|Path,           # full path to dir where input files are\n",
    "        output_dir: str|Path=None,     # full path to dir where to save output files, if different from input_dir\n",
    "        app_in_system_path:bool=False, # whether `art_illumina` is in the system path or not\n",
    "        ):\n",
    "        \"\"\"Initialize the art_illumina instance\"\"\"\n",
    "\n",
    "        # Validate and save paths\n",
    "        path2app = _validate_path(path2app)\n",
    "        if app_in_system_path:\n",
    "            self.app_cmd = 'art_illumina'\n",
    "        elif path2app.is_file():\n",
    "            self.app_cmd = str(path2app.absolute())\n",
    "        else:\n",
    "            raise ValueError(f\"{path2app.name} is not a file, please check the path to the application\")\n",
    "\n",
    "        input_dir = _validate_path(input_dir)\n",
    "        if input_dir.is_dir():\n",
    "            self.input_dir = input_dir\n",
    "        else:\n",
    "            raise ValueError(f\"{input_dir.name} is not a directory, please check the path\")\n",
    "\n",
    "        if output_dir is None: \n",
    "            self.output_dir = input_dir\n",
    "        else:\n",
    "            output_dir = _validate_path(output_dir)\n",
    "            if output_dir.is_dir():\n",
    "                self.output_dir = output_dir\n",
    "            else:\n",
    "                raise ValueError(f\"{output_dir.name} is not a directory, please check the path\")\n",
    "\n",
    "        print(f\"Ready to operate with art: {self.app_cmd}\")\n",
    "        print(f\"Input files from : {self.input_dir.absolute()}\")\n",
    "        print(f\"Output files to :  {self.output_dir.absolute()}\")\n",
    "\n",
    "    def sim_reads(\n",
    "        self,\n",
    "        input_file: str,          # name of the fasta file to use as input\n",
    "        output_seed: str,         # seed to use for the output files\n",
    "        sim_type: str='single',   # type of read simmulation: 'single' or 'paired'\n",
    "        read_length: int=150,     # length of the read in bp\n",
    "        fold: int=10,             # fold\n",
    "        mean_read: int=None,      # mean length of the read for paired reads\n",
    "        std_read: int=None,       # std of the read length, for paired reads\n",
    "        ss: str='HS25',           # quality profile to use for simulation,\n",
    "        overwrite: bool=False,    # overwrite existing output files if true, raise error if false \n",
    "        print_output:bool=True    # if True, prints art ilumina's CLI output\n",
    "        ):\n",
    "        \"\"\"Simulates reads with art_illumina. Output files saved in a separate directory\"\"\"\n",
    "\n",
    "        # validate input file and save in instance\n",
    "        path2inputfile = self.input_dir / input_file\n",
    "        if path2inputfile.is_file(): \n",
    "            self.last_input_file = path2inputfile\n",
    "        else:\n",
    "            raise ValueError(f\"{input_file} is not a file in {self.input_dir}\")\n",
    "\n",
    "        \n",
    "        # validate output seed and save in instance\n",
    "        if not overwrite and len(list(self.output_dir.glob(f\"{output_seed}*\"))) > 0: \n",
    "            raise ValueError(f\"Existing output directory starting with {output_seed}\")\n",
    "        else:\n",
    "            self.last_output_seed = output_seed\n",
    "            self.last_read_output_dir = self.output_dir/self.last_output_seed\n",
    "            os.makedirs(self.last_read_output_dir, exist_ok=True)\n",
    "\n",
    "        # validate quality profile\n",
    "        if ss not in list(self.QUALITY_PROFILES.keys()): \n",
    "            raise ValueError(f\"{ss} is not a built-in profile.\\nPick one of these: {self.QUALITY_PROFILES}\")\n",
    "\n",
    "        # build art_illumina command\n",
    "        if sim_type == 'single':\n",
    "            params = f\"-ss {ss} -l {read_length} -f {fold}\"\n",
    "        elif sim_type == 'paired':\n",
    "            if mean_read is None or std_read is None:\n",
    "                raise ValueError(f\"mean_read and std_read are required for a paired reads simulation\")\n",
    "            else:\n",
    "                params = f\"-ss {ss} -p -l {read_length} -f {fold} -m {mean_read} -s {std_read}\"\n",
    "        else:\n",
    "            raise RuntimeError(f\"{sim_type} in not a type or is not implemented yet\")\n",
    "\n",
    "        p2in = self.last_input_file.absolute()\n",
    "        p2out = (self.output_dir / self.last_output_seed / self.last_output_seed).absolute()\n",
    "\n",
    "        cmd = f\"{self.app_cmd} -i {p2in} {params} -o {p2out}\"\n",
    "        _run(args=shlex.split(cmd), print_output=print_output)\n",
    "\n",
    "    def get_last_output_files(self):\n",
    "        \"\"\"Returns the path to all output files from last simulation\"\"\"\n",
    "        return [f for f in self.last_read_output_dir.iterdir()]\n",
    "\n",
    "    def list_last_output_files(self):\n",
    "        \"\"\"Prints a list of the last simulation's output files\"\"\"\n",
    "        for f in self.get_last_output_files():\n",
    "            print(f.name)\n",
    "\n",
    "    def list_all_input_files(self):\n",
    "        for f in sorted(list(self.input_dir.iterdir())):\n",
    "            if f.suffix == '.fa': print(f.name)\n",
    "\n",
    "    def get_all_output_files(self):\n",
    "        \"\"\"Return a dictionary with k as output file subdirectory and v as list of output files\"\"\"\n",
    "        all_output_files = {}\n",
    "        for d in sorted([p for p in self.output_dir.iterdir() if p.is_dir()]):\n",
    "            files = []\n",
    "            for f in d.iterdir():\n",
    "                files.append(f)\n",
    "            all_output_files[d.name] = files\n",
    "\n",
    "        return all_output_files\n",
    "\n",
    "    def list_all_output_files(self):\n",
    "        all_files = self.get_all_output_files()\n",
    "        for k, v in all_files.items():\n",
    "            print(k)\n",
    "            print('\\n'.join([f\"- {p.name}\" for p in v]))\n",
    "        \n",
    "    def print_last_output_file_excerpts(\n",
    "        self, \n",
    "        suffix: str='fq',        # suffix of the files to explore: 'fq', 'aln', 'sam' \n",
    "        nlines: int=12          # number of lines to print for each file\n",
    "        ):\n",
    "        \"\"\"Print the first lines of all the latest file with given suffix\"\"\"\n",
    "\n",
    "        for p in [p for p in self.get_last_output_files() if p.suffix == f\".{suffix}\"]:\n",
    "            print(f\"{'='*120}\")\n",
    "            print(f\"File Name: {p.name}.\")\n",
    "            print(f\"{'-'*80}\")\n",
    "            with open(p, 'r') as f:\n",
    "                n, lines = 0, []\n",
    "                while True:\n",
    "                    n += 1\n",
    "                    line = f.readline()\n",
    "                    if line == '': break\n",
    "                    elif n > nlines: break\n",
    "                    else:\n",
    "                        lines.append(line)\n",
    "\n",
    "            print(''.join(lines))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Create an instance of `ArtIllumina`\n",
    "2. Run a simulation\n",
    "3. Export output files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create an instance of `ArtIllumina` with:\n",
    "- the path to the application on the local system\n",
    "- the directories for input and output files (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#|eval: false\n",
    "p2art = Path('/bin/art_illumina')\n",
    "assert p2art.exists()\n",
    "p2data = Path('data_dev/ncbi/refsequences/cov')\n",
    "assert p2data.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ready to operate with art: /bin/art_illumina\n",
      "Input files from : /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov\n",
      "Output files to :  /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov\n"
     ]
    }
   ],
   "source": [
    "#|eval: false\n",
    "art = ArtIllumina(\n",
    "    path2app=p2art,\n",
    "    input_dir=p2data,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/art.py#L86){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### ArtIllumina.sim_reads\n",
       "\n",
       ">      ArtIllumina.sim_reads (input_file:str, output_seed:str,\n",
       ">                             sim_type:str='single', read_length:int=150,\n",
       ">                             fold:int=10, mean_read:int=None,\n",
       ">                             std_read:int=None, ss:str='HS25',\n",
       ">                             overwrite:bool=False, print_output:bool=True)\n",
       "\n",
       "*Simulates reads with art_illumina. Output files saved in a separate directory*\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| input_file | str |  | name of the fasta file to use as input |\n",
       "| output_seed | str |  | seed to use for the output files |\n",
       "| sim_type | str | single | type of read simmulation: 'single' or 'paired' |\n",
       "| read_length | int | 150 | length of the read in bp |\n",
       "| fold | int | 10 | fold |\n",
       "| mean_read | int | None | mean length of the read for paired reads |\n",
       "| std_read | int | None | std of the read length, for paired reads |\n",
       "| ss | str | HS25 | quality profile to use for simulation, |\n",
       "| overwrite | bool | False | overwrite existing output files if true, raise error if false |\n",
       "| print_output | bool | True | if True, prints art ilumina's CLI output |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/vtecftwy/metagentools/blob/main/metagentools/art.py#L86){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### ArtIllumina.sim_reads\n",
       "\n",
       ">      ArtIllumina.sim_reads (input_file:str, output_seed:str,\n",
       ">                             sim_type:str='single', read_length:int=150,\n",
       ">                             fold:int=10, mean_read:int=None,\n",
       ">                             std_read:int=None, ss:str='HS25',\n",
       ">                             overwrite:bool=False, print_output:bool=True)\n",
       "\n",
       "*Simulates reads with art_illumina. Output files saved in a separate directory*\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| input_file | str |  | name of the fasta file to use as input |\n",
       "| output_seed | str |  | seed to use for the output files |\n",
       "| sim_type | str | single | type of read simmulation: 'single' or 'paired' |\n",
       "| read_length | int | 150 | length of the read in bp |\n",
       "| fold | int | 10 | fold |\n",
       "| mean_read | int | None | mean length of the read for paired reads |\n",
       "| std_read | int | None | std of the read length, for paired reads |\n",
       "| ss | str | HS25 | quality profile to use for simulation, |\n",
       "| overwrite | bool | False | overwrite existing output files if true, raise error if false |\n",
       "| print_output | bool | True | if True, prints art ilumina's CLI output |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nbdev.show_doc(ArtIllumina.sim_reads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run a single read simulations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Provide an input file and a seed for the names of the output files\n",
    "- Prints out the log messages issued by `art_illumina`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "return code:  0 \n",
      "\n",
      "\n",
      "    ====================ART====================\n",
      "             ART_Illumina (2008-2016)          \n",
      "          Q Version 2.5.8 (June 6, 2016)       \n",
      "     Contact: Weichun Huang <whduke@gmail.com> \n",
      "    -------------------------------------------\n",
      "\n",
      "                  Single-end Simulation\n",
      "\n",
      "Total CPU time used: 0.436844\n",
      "\n",
      "The random seed for the run: 1738405239\n",
      "\n",
      "Parameters used during run\n",
      "\tRead Length:\t150\n",
      "\tGenome masking 'N' cutoff frequency: \t1 in 150\n",
      "\tFold Coverage:            100X\n",
      "\tProfile Type:             Combined\n",
      "\tID Tag:                   \n",
      "\n",
      "Quality Profile(s)\n",
      "\tFirst Read:   HiSeq 2500 Length 150 R1 (built-in profile) \n",
      "\n",
      "Output files\n",
      "\n",
      "  FASTQ Sequence File:\n",
      "\t/home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov/single_1seq_150bp/single_1seq_150bp.fq\n",
      "\n",
      "  ALN Alignment File:\n",
      "\t/home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov/single_1seq_150bp/single_1seq_150bp.aln\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#|eval: false\n",
    "input_fname = 'cov_virus_sequence_one.fa'\n",
    "output_seed = 'single_1seq_150bp'\n",
    "\n",
    "art.sim_reads(\n",
    "    input_file=input_fname,\n",
    "    output_seed=output_seed,\n",
    "    sim_type=\"single\",\n",
    "    read_length=150,\n",
    "    fold=100,\n",
    "    overwrite=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run a paired read simulations with the input file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========================================================================================================================\n",
      "File Name: single_1seq_150bp.fq.\n",
      "--------------------------------------------------------------------------------\n",
      "@2591237:ncbi:1-20100\n",
      "GTACCACAGATGTGCACTTTACGTCAGACATTTTAGACTGTACAGTAGCAACCTTGATACATGGTTTACCTCCAATACCTAACAACTTAATGTTAAGCTTGAAAGCATCAATACTACTCTTAGGAGGCAAAAGCCCCTGGGAGTTCATAT\n",
      "+\n",
      "CCCGGGGGG1GGCGJJGJJGJJGJGJJJJJJJGJJ=GGJGGJJGJJGCCJGJGGGJGCGCC=GGJCGCGJGJGGCC=GGCGGGGGGGGG8GG=GG8GGCCJGCCCGGCCCGG=CGGGGCGGCGGGGGGGGGGGGGGCGGGGCCGGGGGCG\n",
      "@2591237:ncbi:1-20099\n",
      "TACACCCTTTGCCAGCTCGCTATGAGCTGTAGCAACGAGTACCTTAAGTTTTTCCATAGGAACACTAAAAGTTGCTGAAAAGGTGTCGACATAAGCATCAAACATCTTAACAGAAACTTCAGTACTATCTCCAACATCTGATACGAGAGC\n",
      "+\n",
      "=CCG=GGGGGGGGJJGJJJCGJGJJJJJJGGJJJJJJGJJJJCJJGJCGGGJGGGGJGJJG(J=JGGGCG=G=CGGGGG=GGGCG8GGGGGGGC8C=GGCJ8G=CGGGGGGGGG=GGGG=1G8G==GCGGGGGCGGGGGGGGCCGCCCGC\n",
      "@2591237:ncbi:1-20098\n",
      "ATGTCCTGCCTGTCAAGACCCAGAGATTGGACCTGAGCATAGTGTTGCAGATTATCACAACCACTCAAACATTGAAACTCGACTCCGCAAGGGAGGTAGGACTAGATGTTTTGGAGGCTGTGTGTTTGCCTATGTCGGCTGCTATAACAA\n",
      "+\n",
      "CCC1CGG1GGGGGJJJJCGJJ1JGJGJGJJGJGGJGJGGJJGJGJGJJGCGJCJ=JJGCGGGCJG1CGCGC=GCGGGCGCGG=GGCGGGGCG8GCGGGCGCCG=GCGGGGG(GGGCGGGGG=CGGGCGCGGC8CGGGCGCGCCGGGGGGG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#|eval: false\n",
    "art.print_last_output_file_excerpts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "return code:  0 \n",
      "\n",
      "\n",
      "    ====================ART====================\n",
      "             ART_Illumina (2008-2016)          \n",
      "          Q Version 2.5.8 (June 6, 2016)       \n",
      "     Contact: Weichun Huang <whduke@gmail.com> \n",
      "    -------------------------------------------\n",
      "\n",
      "                  Paired-end sequencing simulation\n",
      "\n",
      "Total CPU time used: 0.437712\n",
      "\n",
      "The random seed for the run: 1738405243\n",
      "\n",
      "Parameters used during run\n",
      "\tRead Length:\t150\n",
      "\tGenome masking 'N' cutoff frequency: \t1 in 150\n",
      "\tFold Coverage:            100X\n",
      "\tMean Fragment Length:     200\n",
      "\tStandard Deviation:       10\n",
      "\tProfile Type:             Combined\n",
      "\tID Tag:                   \n",
      "\n",
      "Quality Profile(s)\n",
      "\tFirst Read:   HiSeq 2500 Length 150 R1 (built-in profile) \n",
      "\tFirst Read:   HiSeq 2500 Length 150 R2 (built-in profile) \n",
      "\n",
      "Output files\n",
      "\n",
      "  FASTQ Sequence Files:\n",
      "\t the 1st reads: /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov/paired_1seq_150bp/paired_1seq_150bp1.fq\n",
      "\t the 2nd reads: /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov/paired_1seq_150bp/paired_1seq_150bp2.fq\n",
      "\n",
      "  ALN Alignment Files:\n",
      "\t the 1st reads: /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov/paired_1seq_150bp/paired_1seq_150bp1.aln\n",
      "\t the 2nd reads: /home/vtec/projects/bio/metagentools/nbs-dev/data_dev/ncbi/refsequences/cov/paired_1seq_150bp/paired_1seq_150bp2.aln\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#|eval: false\n",
    "input_fname = 'cov_virus_sequence_one.fa'\n",
    "\n",
    "art.sim_reads(\n",
    "    input_file=input_fname,\n",
    "    output_seed='paired_1seq_150bp',\n",
    "    sim_type=\"paired\",\n",
    "    read_length=150,\n",
    "    fold=100,\n",
    "    mean_read=200,\n",
    "    std_read=10,\n",
    "    overwrite=True\n",
    ")   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========================================================================================================================\n",
      "File Name: paired_1seq_150bp2.fq.\n",
      "--------------------------------------------------------------------------------\n",
      "@2591237:ncbi:1-20100/2\n",
      "TTATAGCAGCCGACATAGGCAAACACACAGCCTCCAAAACATCTAGTCCTACCTCCCTTGCGGAGTCGAGTTTCAATGTTTGAGTGGTTGTGATAATCTGCAACACTATGCTCAGGTCCAATCTCTGGGTCTTGACAGGCAGGACATGGC\n",
      "+\n",
      "CCCGGGGGCGGGGGJJJJJGJJJJG8JJ=GJJCGGJCJ1GGCJJGCGGJJJJGGCJGGCGJJ=JCGGG=GGG(C=GCCGC=GGGGCGCGGGGGGGGGG=GCGCCJJJCGGGGCCGCGGCCGGCCGGGCC8CGGGCGC=GGGCGCGCCCCC\n",
      "@2591237:ncbi:1-20098/2\n",
      "ATCATTACCGGTCTTCATCCAACACAGGCACCTACACACCTCAGCGTTGACACAAAATTTAAGACTGAGGGACTATGTGTTGACATACCAGGCATACCAAAGGACATGACCTACCGTAGACTCATCTCTATGATGGGTTTTAAAATGAAT\n",
      "+\n",
      "=CCGG=G1GGGGGCGJJGGJJJJ8JJJJ=JCGJJGJJGJJJJCJJGGGJJGGGGG=CJGCGGGCCJJ8CG8J=CGGGGCCGGGGGCGCCCGGGGGCGGCG=CJCJCJ=C=GGGCGGCGGGG=CGGGGGGCGGGCGGGGGGGCGCGCGGGC\n",
      "@2591237:ncbi:1-20096/2\n",
      "CGGTACTAGACATACCTATCAGCTTCGTGCAAGATCAGTTTCACCAAAACTTTTCATCAGACAAGAGGAAGTTCACCAAGAGCTCTACTCACCGCTTTTTCTCATTGTTGCTGCTCTAGTATTTATAATACTTTGCTTCACCATTAAGAG\n",
      "+\n",
      "CCCGGGGGGG=GC1JJJJJJJJ1JJJCGCJJJGCJGJCG(JGGJGJGJGJGGGCCJJCCJJGG8JCGGCGG=G=J8JCGG=8GCCGCGCGGG=C8GCGGG=CJJCJJCGGG1GGG=GG=GCGGC(GCCCGGG=GCGCCGCCGCGCG=GC=\n",
      "\n",
      "========================================================================================================================\n",
      "File Name: paired_1seq_150bp1.fq.\n",
      "--------------------------------------------------------------------------------\n",
      "@2591237:ncbi:1-20100/1\n",
      "TGAAGGACCTACTACATGTGGGTACCTACCTACTAATGCTGTAGTGAAAATGCCATGTCCTGCCTGTCAAGACCCAGAGATTGGACCTGAGCATAGTGTTGCAGATTATCACAACCACTCAAACATTGAAACTCGACTCCGCAAGGGAGG\n",
      "+\n",
      "CC1CGGGGGGGGGJJJCGJGGJGJJJJGGCGJ=J1JGJJJJJGGGG1GJJGJGJGJGGCGGGGGJGGJJGGGJCCCGCGCGG=GGGCGGGGGGGGG=CG1JCGGC(GGCCGCC8GGGGGGCGGGCGGCCGCGCCCGCGGCCCGGGGC8GC\n",
      "@2591237:ncbi:1-20098/1\n",
      "TAGCTTCTTCGCGGGTGATAAACATATTAGCGTAACCATTGACTTGGTAATTCATTTTAAAACCCATCATAGAGATGAGTCTACGGTAGGTCATGTCCTTTGGTATGCCTGGTATGTCAACACATAGTCCCTCAGTTTAAAATTTTGTGT\n",
      "+\n",
      "1CCGGGGGGCGGGJJJGJGJJGGJJGJGGG(GJJJGJJCCJJJCJGJJGCGJGJGG(GGGGGGCJGGGJGGGCCGGGGGGGGGCGGCGGGGCGGGCG(GGJCCGGGGCGG=1GGCGCG=GCGCGGG=CGGGCCC1G(G(CG=GCGGGCGG\n",
      "@2591237:ncbi:1-20096/1\n",
      "GAATAGCAGAAAGGCTAAAAAGCACAAATAGAAGTCAATTAAAGTGAGCTCATTCATTCTGTCTTTCTCTTAATGGTGAAGCAAAGTATTATAAATACTAGAGCAGCAACAATGAGAAAAAGCGGTGAGTAGAGCTCTTGGTGAACTTCC\n",
      "+\n",
      "CCCGGGCGGGGGGJJGJJJJGJGJJGJGJ8JJGGJJCJGG=JJJCJC(JJGJJJGGJJGG1GCGGGJGCCJCGGCJGCJGCGCCCGGCGGGGGGG8GGGGJGGGG1GCCGGGCGGGCCGGCC=G=GGGGCCGG=CG=CGCGGGCGGGGGC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#|eval: false\n",
    "art.print_last_output_file_excerpts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "paired_1seq_150bp\n",
      "- paired_1seq_150bp2.aln\n",
      "- paired_1seq_150bp2.fq\n",
      "- paired_1seq_150bp1.fq\n",
      "- paired_1seq_150bp1.aln\n",
      "single_1seq_150bp\n",
      "- single_1seq_150bp.fq\n",
      "- single_1seq_150bp.aln\n"
     ]
    }
   ],
   "source": [
    "#|eval: false\n",
    "art.list_all_output_files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "26b205197a934fdeabb71e65ac11acba": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "514ad0bfcabf4df580a9a872af814af9": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    },
    "55646397fc9349d3af9e98b1f2b26f5d": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "70f6be4247664b708b662c34e7abe3ee": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "7179c6cc207941648c348b1bf10cb87f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "LabelModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "LabelModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "LabelView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_70f6be4247664b708b662c34e7abe3ee",
      "placeholder": "​",
      "style": "IPY_MODEL_bdde467d943148ce9bb6355fd7582a5c",
      "value": "0.078 MB of 0.078 MB uploaded (0.020 MB deduped)\r"
     }
    },
    "7849f255e99b4853bdba7f8badf1054a": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "VBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "VBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "VBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_7179c6cc207941648c348b1bf10cb87f",
       "IPY_MODEL_e0819a1ddcc64c08a748a2fd88350f09"
      ],
      "layout": "IPY_MODEL_26b205197a934fdeabb71e65ac11acba"
     }
    },
    "bdde467d943148ce9bb6355fd7582a5c": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "e0819a1ddcc64c08a748a2fd88350f09": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_55646397fc9349d3af9e98b1f2b26f5d",
      "max": 1,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_514ad0bfcabf4df580a9a872af814af9",
      "value": 1
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
